diff --git a/3rdparty/pcre-orig/config.sub b/3rdparty/pcre-workdir/config.sub
index c874b7a..40d8304 100755
--- a/3rdparty/pcre-orig/config.sub
+++ b/3rdparty/pcre-workdir/config.sub
@@ -1695,7 +1695,7 @@ case $os in
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
-	     | os9* | macos* | osx* | ios* \
+	     | os9* | macos* | osx* | ios* | vinix* \
 	     | mpw* | magic* | mmixware* | mon960* | lnews* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* | twizzler* \
