From 957cf55cacc1bf60c82f78b9064c158f0eb9b4a3 Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Sat, 14 May 2022 18:44:44 +0200
Subject: [PATCH] Vinix specific changes

---
 configure.ac | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/configure.ac b/configure.ac
index 7c63ed7..6ab1502 100644
--- a/configure.ac
+++ b/configure.ac
@@ -72,12 +72,12 @@ AM_CONDITIONAL(BUILDING_FROM_GIT, test x$from_git = xyes)
 dnl Checks for pkg-config and gettext when building from git.
 
 if test x$from_git = xyes; then
-	if test ! -f $(aclocal --print-ac-dir)/pkg.m4; then
-		AC_MSG_ERROR([
-  *** The pkg.m4 macros are missing.
-  *** The pkg-config package needs to be installed when building from git.
-  *** After fixing this problem, rerun ./autogen.sh.])
-	fi
+dnl	if test ! -f $(aclocal --print-ac-dir)/pkg.m4; then
+dnl		AC_MSG_ERROR([
+dnl  *** The pkg.m4 macros are missing.
+dnl  *** The pkg-config package needs to be installed when building from git.
+dnl  *** After fixing this problem, rerun ./autogen.sh.])
+dnl	fi
 	if test "$ac_cv_path_MSGFMT" = ":"; then
 		AC_MSG_ERROR([
   *** The msgfmt program is missing.
-- 
2.35.1

