diff --git a/3rdparty/vc-orig/v.c b/3rdparty/vc-workdir/v.c
index 192ee6c..a3a50da 100644
--- a/3rdparty/vc-orig/v.c
+++ b/3rdparty/vc-workdir/v.c
@@ -625,6 +625,7 @@ typedef int (*qsort_callback_func)(const void*, const void*);
 #include <stdio.h>  // TODO remove all these includes, define all function signatures and types manually
 #include <stdlib.h>
 #include <string.h>
+#include <sys/wait.h>
 
 #if defined(_WIN32) || defined(__CYGWIN__)
 	#define VV_EXPORTED_SYMBOL extern __declspec(dllexport)
