diff -urN --no-dereference xorg-proto-clean/include/X11/Xfuncs.h xorg-proto-workdir/include/X11/Xfuncs.h
--- xorg-proto-clean/include/X11/Xfuncs.h
+++ xorg-proto-workdir/include/X11/Xfuncs.h
@@ -44,7 +44,7 @@
 #    define bcmp(b1,b2,len) memcmp(b1, b2, len)
 #   else
 #    include <string.h>
-#    if defined(__SCO__) || defined(__sun) || defined(__UNIXWARE__) || defined(__CYGWIN__) || defined(_AIX) || defined(__APPLE__)
+#    if defined(__SCO__) || defined(__sun) || defined(__UNIXWARE__) || defined(__CYGWIN__) || defined(_AIX) || defined(__APPLE__) || defined(__vinix__)
 #     include <strings.h>
 #    endif
 #    define _XFUNCS_H_INCLUDED_STRING_H
diff -urN --no-dereference xorg-proto-clean/include/X11/Xos.h xorg-proto-workdir/include/X11/Xos.h
--- xorg-proto-clean/include/X11/Xos.h
+++ xorg-proto-workdir/include/X11/Xos.h
@@ -60,7 +60,7 @@
  */
 
 # include <string.h>
-# if defined(__SCO__) || defined(__UNIXWARE__) || defined(__sun) || defined(__CYGWIN__) || defined(_AIX) || defined(__APPLE__) || defined(__FreeBSD__)
+# if defined(__SCO__) || defined(__UNIXWARE__) || defined(__sun) || defined(__CYGWIN__) || defined(_AIX) || defined(__APPLE__) || defined(__FreeBSD__) || defined(__vinix__)
 #  include <strings.h>
 # else
 #  ifndef index
diff -urN --no-dereference xorg-proto-clean/include/X11/Xos_r.h xorg-proto-workdir/include/X11/Xos_r.h
--- xorg-proto-clean/include/X11/Xos_r.h
+++ xorg-proto-workdir/include/X11/Xos_r.h
@@ -318,7 +318,7 @@
   (_Xos_processUnlock), \
   (p).pwp )
 
-#elif !defined(_POSIX_THREAD_SAFE_FUNCTIONS) && !defined(__APPLE__)
+#elif !defined(_POSIX_THREAD_SAFE_FUNCTIONS) && !defined(__APPLE__) && !defined(__vinix__)
 # define X_NEEDS_PWPARAMS
 typedef struct {
   struct passwd pws;
