diff --git xf86-input-keyboard-clean/configure.ac xf86-input-keyboard-workdir/configure.ac
index 0f71ce3..9d7adf9 100644
--- xf86-input-keyboard-clean/configure.ac
+++ xf86-input-keyboard-workdir/configure.ac
@@ -57,6 +57,10 @@ AC_SUBST(inputdir)
 
 # The keyboard driver code is O/S specific
 case $host_os in
+  vinix*)
+    IS_VINIX="yes"
+    ;;
+
   linux*)
     AC_MSG_ERROR([This is not the keyboard driver you are looking for. Use evdev or libinput.])
     ;;
@@ -95,6 +99,7 @@ case $host_os in
 esac
 AC_SUBST([OS_FLAGS])
 
+AM_CONDITIONAL(VINIX, [test "x$IS_VINIX" = xyes])
 AM_CONDITIONAL(BSD, [test "x$IS_BSD" = xyes])
 AM_CONDITIONAL(SOLARIS, [test "x$IS_SOLARIS" = xyes])
 AM_CONDITIONAL(HURD, [test "x$IS_HURD" = xyes])
diff --git xf86-input-keyboard-clean/src/Makefile.am xf86-input-keyboard-workdir/src/Makefile.am
index 52f5f4d..643e707 100644
--- xf86-input-keyboard-clean/src/Makefile.am
+++ xf86-input-keyboard-workdir/src/Makefile.am
@@ -26,10 +26,15 @@ kbd_drv_la_SOURCES = kbd.c xf86OSKbd.h xf86Keymap.h atKeynames.h
 kbd_drv_la_LIBADD = $(XORG_LIBS)
 kbd_drv_ladir = @inputdir@
 
+VINIX_SRCS = vinix_kbd.c at_scancode.c
 BSD_SRCS = bsd_KbdMap.c bsd_kbd.c bsd_kbd.h at_scancode.c
 HURD_SRCS = hurd_kbd.c at_scancode.c
 SOLARIS_SRCS = sun_kbd.c sun_kbd.h sun_kbdMap.c
 
+if VINIX
+kbd_drv_la_SOURCES += $(VINIX_SRCS)
+endif
+
 if BSD
 kbd_drv_la_SOURCES += $(BSD_SRCS)
 endif
@@ -42,4 +47,4 @@ if HURD
 kbd_drv_la_SOURCES += $(HURD_SRCS)
 endif
 
-EXTRA_DIST = $(BSD_SRCS) $(HURD_SRCS) $(SOLARIS_SRCS)
+EXTRA_DIST = $(VINIX_SRCS) $(BSD_SRCS) $(HURD_SRCS) $(SOLARIS_SRCS)
diff --git xf86-input-keyboard-workdir/src/vinix_kbd.c xf86-input-keyboard-workdir/src/vinix_kbd.c
new file mode 100644
index 0000000..b3d3d19
--- /dev/null
+++ xf86-input-keyboard-workdir/src/vinix_kbd.c
@@ -0,0 +1,31 @@
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <xorg-server.h>
+#include <X11/X.h>
+
+#include "compiler.h"
+
+#include "xf86.h"
+#include "xf86Priv.h"
+#include "xf86_OSlib.h"
+
+#include "xf86Xinput.h"
+#include "xf86OSKbd.h"
+#include "atKeynames.h"
+#include "xf86Keymap.h"
+
+#include <stdio.h>
+#include <errno.h>
+#include <sys/time.h>
+#include <sys/file.h>
+#include <assert.h>
+#include <sys/ioctl.h>
+
+Bool
+xf86OSKbdPreInit(InputInfoPtr pInfo)
+{
+    return FALSE;
+}
