diff -urN --no-dereference xf86-input-keyboard-clean/configure xf86-input-keyboard-workdir/configure
--- xf86-input-keyboard-clean/configure	2024-11-14 21:26:12.000000000 +0100
+++ xf86-input-keyboard-workdir/configure	2025-04-26 23:43:59.711931244 +0200
@@ -664,6 +664,8 @@
 SOLARIS_TRUE
 BSD_FALSE
 BSD_TRUE
+VINIX_FALSE
+VINIX_TRUE
 OS_FLAGS
 inputdir
 XORG_LIBS
@@ -19840,6 +19842,10 @@
 
 # The keyboard driver code is O/S specific
 case $host_os in
+  vinix*)
+    IS_VINIX="yes"
+    ;;
+
   linux*)
     as_fn_error $? "This is not the keyboard driver you are looking for. Use evdev or libinput." "$LINENO" 5
     ;;
@@ -19878,6 +19884,14 @@
 esac
 
 
+ if test "x$IS_VINIX" = xyes; then
+  VINIX_TRUE=
+  VINIX_FALSE='#'
+else
+  VINIX_TRUE='#'
+  VINIX_FALSE=
+fi
+
  if test "x$IS_BSD" = xyes; then
   BSD_TRUE=
   BSD_FALSE='#'
@@ -20053,6 +20067,10 @@
   as_fn_error $? "conditional \"am__fastdepCC\" was never defined.
 Usually this means the macro was only invoked conditionally." "$LINENO" 5
 fi
+if test -z "${VINIX_TRUE}" && test -z "${VINIX_FALSE}"; then
+  as_fn_error $? "conditional \"VINIX\" was never defined.
+Usually this means the macro was only invoked conditionally." "$LINENO" 5
+fi
 if test -z "${BSD_TRUE}" && test -z "${BSD_FALSE}"; then
   as_fn_error $? "conditional \"BSD\" was never defined.
 Usually this means the macro was only invoked conditionally." "$LINENO" 5
diff -urN --no-dereference xf86-input-keyboard-clean/src/Makefile.in xf86-input-keyboard-workdir/src/Makefile.in
--- xf86-input-keyboard-clean/src/Makefile.in	2024-11-14 21:26:13.000000000 +0100
+++ xf86-input-keyboard-workdir/src/Makefile.in	2025-04-27 00:21:15.624904482 +0200
@@ -113,6 +113,7 @@
 @BSD_TRUE@am__append_1 = $(BSD_SRCS)
 @SOLARIS_TRUE@am__append_2 = $(SOLARIS_SRCS)
 @HURD_TRUE@am__append_3 = $(HURD_SRCS)
+@VINIX_TRUE@am__append_4 = $(VINIX_SRCS)
 subdir = src
 ACLOCAL_M4 = $(top_srcdir)/aclocal.m4
 am__aclocal_m4_deps = $(top_srcdir)/configure.ac
@@ -155,15 +156,17 @@
 kbd_drv_la_DEPENDENCIES = $(am__DEPENDENCIES_1)
 am__kbd_drv_la_SOURCES_DIST = kbd.c xf86OSKbd.h xf86Keymap.h \
 	atKeynames.h bsd_KbdMap.c bsd_kbd.c bsd_kbd.h at_scancode.c \
-	sun_kbd.c sun_kbd.h sun_kbdMap.c hurd_kbd.c
+	sun_kbd.c sun_kbd.h sun_kbdMap.c hurd_kbd.c vinix_kbd.c
 am__objects_1 = bsd_KbdMap.lo bsd_kbd.lo at_scancode.lo
 @BSD_TRUE@am__objects_2 = $(am__objects_1)
 am__objects_3 = sun_kbd.lo sun_kbdMap.lo
 @SOLARIS_TRUE@am__objects_4 = $(am__objects_3)
 am__objects_5 = hurd_kbd.lo at_scancode.lo
 @HURD_TRUE@am__objects_6 = $(am__objects_5)
+am__objects_7 = vinix_kbd.lo at_scancode.lo
+@VINIX_TRUE@am__objects_8 = $(am__objects_7)
 am_kbd_drv_la_OBJECTS = kbd.lo $(am__objects_2) $(am__objects_4) \
-	$(am__objects_6)
+	$(am__objects_6) $(am__objects_8)
 kbd_drv_la_OBJECTS = $(am_kbd_drv_la_OBJECTS)
 AM_V_lt = $(am__v_lt_@AM_V@)
 am__v_lt_ = $(am__v_lt_@AM_DEFAULT_V@)
@@ -189,7 +192,7 @@
 am__maybe_remake_depfiles = depfiles
 am__depfiles_remade = ./$(DEPDIR)/at_scancode.Plo \
 	./$(DEPDIR)/bsd_KbdMap.Plo ./$(DEPDIR)/bsd_kbd.Plo \
-	./$(DEPDIR)/hurd_kbd.Plo ./$(DEPDIR)/kbd.Plo \
+	./$(DEPDIR)/hurd_kbd.Plo ./$(DEPDIR)/vinix_kbd.Plo ./$(DEPDIR)/kbd.Plo \
 	./$(DEPDIR)/sun_kbd.Plo ./$(DEPDIR)/sun_kbdMap.Plo
 am__mv = mv -f
 COMPILE = $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) \
@@ -388,13 +391,14 @@
 kbd_drv_la_LTLIBRARIES = kbd_drv.la
 kbd_drv_la_LDFLAGS = -avoid-version -module
 kbd_drv_la_SOURCES = kbd.c xf86OSKbd.h xf86Keymap.h atKeynames.h \
-	$(am__append_1) $(am__append_2) $(am__append_3)
+	$(am__append_1) $(am__append_2) $(am__append_3) $(am__append_4)
 kbd_drv_la_LIBADD = $(XORG_LIBS)
 kbd_drv_ladir = @inputdir@
+VINIX_SRCS = vinix_kbd.c at_scancode.c
 BSD_SRCS = bsd_KbdMap.c bsd_kbd.c bsd_kbd.h at_scancode.c
 HURD_SRCS = hurd_kbd.c at_scancode.c
 SOLARIS_SRCS = sun_kbd.c sun_kbd.h sun_kbdMap.c
-EXTRA_DIST = $(BSD_SRCS) $(HURD_SRCS) $(SOLARIS_SRCS)
+EXTRA_DIST = $(VINIX_SRCS) $(BSD_SRCS) $(HURD_SRCS) $(SOLARIS_SRCS)
 all: all-am
 
 .SUFFIXES:
@@ -475,6 +479,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/bsd_KbdMap.Plo@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/bsd_kbd.Plo@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/hurd_kbd.Plo@am__quote@ # am--include-marker
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/vinix_kbd.Plo@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/kbd.Plo@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/sun_kbd.Plo@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/sun_kbdMap.Plo@am__quote@ # am--include-marker
@@ -643,6 +648,7 @@
 	-rm -f ./$(DEPDIR)/bsd_KbdMap.Plo
 	-rm -f ./$(DEPDIR)/bsd_kbd.Plo
 	-rm -f ./$(DEPDIR)/hurd_kbd.Plo
+	-rm -f ./$(DEPDIR)/vinix_kbd.Plo
 	-rm -f ./$(DEPDIR)/kbd.Plo
 	-rm -f ./$(DEPDIR)/sun_kbd.Plo
 	-rm -f ./$(DEPDIR)/sun_kbdMap.Plo
@@ -695,6 +701,7 @@
 	-rm -f ./$(DEPDIR)/bsd_KbdMap.Plo
 	-rm -f ./$(DEPDIR)/bsd_kbd.Plo
 	-rm -f ./$(DEPDIR)/hurd_kbd.Plo
+	-rm -f ./$(DEPDIR)/vinix_kbd.Plo
 	-rm -f ./$(DEPDIR)/kbd.Plo
 	-rm -f ./$(DEPDIR)/sun_kbd.Plo
 	-rm -f ./$(DEPDIR)/sun_kbdMap.Plo
diff -urN --no-dereference xf86-input-keyboard-clean/src/vinix_kbd.c xf86-input-keyboard-workdir/src/vinix_kbd.c
--- xf86-input-keyboard-clean/src/vinix_kbd.c	1970-01-01 01:00:00.000000000 +0100
+++ xf86-input-keyboard-workdir/src/vinix_kbd.c	2025-04-26 16:38:34.818112788 +0200
@@ -0,0 +1,31 @@
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <xorg-server.h>
+#include <X11/X.h>
+
+#include "compiler.h"
+
+#include "xf86.h"
+#include "xf86Priv.h"
+#include "xf86_OSlib.h"
+
+#include "xf86Xinput.h"
+#include "xf86OSKbd.h"
+#include "atKeynames.h"
+#include "xf86Keymap.h"
+
+#include <stdio.h>
+#include <errno.h>
+#include <sys/time.h>
+#include <sys/file.h>
+#include <assert.h>
+#include <sys/ioctl.h>
+
+Bool
+xf86OSKbdPreInit(InputInfoPtr pInfo)
+{
+    return FALSE;
+}
