From 9d99a635c00c20d30f44ccb63f564ad574eb057b Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Sun, 23 Jan 2022 04:23:08 +0100
Subject: [PATCH] Vinix specific changes

---
 configure.ac    | 8 ++++----
 src/Makefile.am | 2 +-
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/configure.ac b/configure.ac
index 4eb7fae..b5fd8ce 100644
--- a/configure.ac
+++ b/configure.ac
@@ -54,12 +54,12 @@ AC_SUBST(moduledir)
 
 
 # Store the list of server defined optional extensions in REQUIRED_MODULES
-XORG_DRIVER_CHECK_EXT(RANDR, randrproto)
-XORG_DRIVER_CHECK_EXT(RENDER, renderproto)
-XORG_DRIVER_CHECK_EXT(XV, videoproto)
+#XORG_DRIVER_CHECK_EXT(RANDR, randrproto)
+#XORG_DRIVER_CHECK_EXT(RENDER, renderproto)
+#XORG_DRIVER_CHECK_EXT(XV, videoproto)
 
 if test "x$DGA" = xyes; then
-	XORG_DRIVER_CHECK_EXT(XFreeXDGA, xf86dgaproto)
+	#XORG_DRIVER_CHECK_EXT(XFreeXDGA, xf86dgaproto)
 	AC_DEFINE(USE_DGA, 1, [Support DGA extension])
 fi
 AC_SUBST([DGA])
diff --git a/src/Makefile.am b/src/Makefile.am
index da1dd9a..0516a04 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -28,7 +28,7 @@
 AM_CFLAGS = $(XORG_CFLAGS) $(PCIACCESS_CFLAGS)
 
 dummy_drv_la_LTLIBRARIES = dummy_drv.la
-dummy_drv_la_LDFLAGS = -module -avoid-version
+dummy_drv_la_LDFLAGS = -module -avoid-version -R@moduledir@ -L$(SYSROOT)@moduledir@ -lfb
 dummy_drv_la_LIBADD = $(XORG_LIBS)
 dummy_drv_ladir = @moduledir@/drivers
 
-- 
2.34.1

