diff --git git-clean/config.mak.uname git-workdir/config.mak.uname
index a7607a5..cce83ee 100644
--- git-clean/config.mak.uname
+++ git-workdir/config.mak.uname
@@ -4,12 +4,12 @@
 # Microsoft's Safe Exception Handling in libraries (such as zlib).
 # Typically required for VS2013+/32-bit compilation on Vista+ versions.
 
-uname_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
-uname_M := $(shell sh -c 'uname -m 2>/dev/null || echo not')
-uname_O := $(shell sh -c 'uname -o 2>/dev/null || echo not')
-uname_R := $(shell sh -c 'uname -r 2>/dev/null || echo not')
-uname_P := $(shell sh -c 'uname -p 2>/dev/null || echo not')
-uname_V := $(shell sh -c 'uname -v 2>/dev/null || echo not')
+uname_S := Vinix
+uname_M := x86_64
+uname_O := Vinix
+uname_R := 0.0.0
+uname_P := unknown
+uname_V := 0.0.0
 
 ifneq ($(findstring MINGW,$(uname_S)),)
 	uname_S := MINGW
