--- a/jinx
+++ b/jinx
@@ -492,10 +492,18 @@ do_git_fetch() {
     [ -d "${source_dir}" ] && return
     if ! [ -z "${commit}" ]; then
         if [ "${shallow}" = "no" ]; then
-            git clone "${git_url}" --revision="${commit}" "${base_dir}"/sources/${name}
+            git clone "${git_url}" "${base_dir}"/sources/${name}
+            git -C "${base_dir}"/sources/${name} \
+                -c advice.detachedHead=false checkout --detach "${commit}"
         else
-            git clone "${git_url}" --revision="${commit}" --depth=1 "${base_dir}"/sources/${name}
-            git -C "${base_dir}"/sources/${name} fetch --tags
+            mkdir -p "${base_dir}"/sources/${name}
+            git -C "${base_dir}"/sources/${name} init
+            git -C "${base_dir}"/sources/${name} remote add origin "${git_url}"
+            git -C "${base_dir}"/sources/${name} fetch --depth=1 origin "${commit}"
+            git -C "${base_dir}"/sources/${name} \
+                -c advice.detachedHead=false checkout --detach FETCH_HEAD
+            git -C "${base_dir}"/sources/${name} fetch --no-recurse-submodules origin \
+                '+refs/tags/*:refs/tags/*'
         fi
     else
         if [ "${shallow}" = "no" ]; then
