diff --git gdbm-clean/tools/gdbmapp.h gdbm-workdir/tools/gdbmapp.h
index 3ea826b..e20253a 100644
--- gdbm-clean/tools/gdbmapp.h
+++ gdbm-workdir/tools/gdbmapp.h
@@ -16,6 +16,7 @@
 
 #include <stdlib.h>
 #include <stdarg.h>
+#include <sys/types.h>
 #include "gettext.h"
 #ifdef HAVE_LOCALE_H
 # include <locale.h>
diff --git gdbm-clean/tools/var.c gdbm-workdir/tools/var.c
index 2cf922e..6aed417 100644
--- gdbm-clean/tools/var.c
+++ gdbm-workdir/tools/var.c
@@ -14,6 +14,7 @@
    You should have received a copy of the GNU General Public License
    along with GDBM. If not, see <http://www.gnu.org/licenses/>.    */
 
+#include <strings.h>
 #include "gdbmtool.h"
 
 #define VARF_DFL    0x00   /* Default flags -- everything disabled */
