diff --git a/3rdparty/limine-orig/stage23/lib/term.s2.c b/3rdparty/limine-workdir/stage23/lib/term.s2.c
index 58a8204..2a23519 100644
--- a/3rdparty/limine-orig/stage23/lib/term.s2.c
+++ b/3rdparty/limine-workdir/stage23/lib/term.s2.c
@@ -853,6 +853,7 @@ void term_putchar(uint8_t c) {
     get_cursor_pos(&x, &y);
 
     switch (c) {
+        case 0x00:
         case 0x7f:
             return;
         case 0x9b:
