diff --git a/3rdparty/binutils-orig/bfd/config.bfd b/3rdparty/binutils-workdir/bfd/config.bfd
index 30087e3..9ca3863 100644
--- a/3rdparty/binutils-orig/bfd/config.bfd
+++ b/3rdparty/binutils-workdir/bfd/config.bfd
@@ -719,6 +719,11 @@ case "${targ}" in
     targ_defvec=i386_elf32_vec
     targ_selvecs="iamcu_elf32_vec i386_pe_vec i386_pei_vec"
     ;;
+  x86_64-*-vinix*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
   i[3-7]86-*-interix*)
     targ_defvec=i386_pei_vec
     targ_selvecs="i386_pe_vec"
diff --git a/3rdparty/binutils-orig/config.sub b/3rdparty/binutils-workdir/config.sub
index 7384e91..6905e3f 100755
--- a/3rdparty/binutils-orig/config.sub
+++ b/3rdparty/binutils-workdir/config.sub
@@ -1704,7 +1704,7 @@ case $os in
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
-	     | os9* | macos* | osx* | ios* \
+	     | os9* | macos* | osx* | ios* | vinix* \
 	     | mpw* | magic* | mmixware* | mon960* | lnews* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* | twizzler* \
diff --git a/3rdparty/binutils-orig/gas/configure.tgt b/3rdparty/binutils-workdir/gas/configure.tgt
index 338892a..0623579 100644
--- a/3rdparty/binutils-orig/gas/configure.tgt
+++ b/3rdparty/binutils-workdir/gas/configure.tgt
@@ -221,6 +221,7 @@ case ${generic_target} in
   i386-*-beos*)				fmt=elf ;;
   i386-*-elfiamcu)			fmt=elf arch=iamcu ;;
   i386-*-elf*)				fmt=elf ;;
+  i386-*-vinix*)				fmt=elf ;;
   i386-*-fuchsia*)			fmt=elf ;;
   i386-*-genode*)			fmt=elf ;;
   i386-*-bsd*)				fmt=aout em=386bsd ;;
diff --git a/3rdparty/binutils-orig/ld/configure.tgt b/3rdparty/binutils-workdir/ld/configure.tgt
index 6205d7c..dba5702 100644
--- a/3rdparty/binutils-orig/ld/configure.tgt
+++ b/3rdparty/binutils-workdir/ld/configure.tgt
@@ -318,6 +318,10 @@ i[3-7]86-*-bsd386)	targ_emul=i386bsd
 i[3-7]86-*-bsdi*)	targ_emul=i386bsd
 			targ_extra_ofiles=
 			;;
+x86_64-*-vinix*)
+			targ_emul=elf_x86_64
+			targ_extra_emuls=elf_i386
+			;;
 i[3-7]86-*-linux-*)	targ_emul=elf_i386
 			targ_extra_emuls="elf_iamcu"
 			targ64_extra_emuls="elf_x86_64 elf32_x86_64 elf_l1om elf_k1om"
diff --git a/3rdparty/binutils-orig/libiberty/rust-demangle.c b/3rdparty/binutils-workdir/libiberty/rust-demangle.c
index df09b7b..8547b2a 100644
--- a/3rdparty/binutils-orig/libiberty/rust-demangle.c
+++ b/3rdparty/binutils-workdir/libiberty/rust-demangle.c
@@ -32,6 +32,8 @@ If not, see <http://www.gnu.org/licenses/>.  */
 #include "config.h"
 #endif
 
+#define uint unsigned int
+
 #include "safe-ctype.h"
 
 #include <inttypes.h>
