diff --git a/ports/binutils-orig/bfd/config.bfd b/ports/binutils-workdir/bfd/config.bfd
index e7b55b0..e3d1e2a 100644
--- a/ports/binutils-orig/bfd/config.bfd
+++ b/ports/binutils-workdir/bfd/config.bfd
@@ -724,6 +724,11 @@ case "${targ}" in
     targ_defvec=i386_elf32_vec
     targ_selvecs="iamcu_elf32_vec i386_pe_vec i386_pei_vec"
     ;;
+  x86_64-*-vos*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
   i[3-7]86-*-interix*)
     targ_defvec=i386_pei_vec
     targ_selvecs="i386_pe_vec"
diff --git a/ports/binutils-orig/config.sub b/ports/binutils-workdir/config.sub
index b0f8492..60f9513 100755
--- a/ports/binutils-orig/config.sub
+++ b/ports/binutils-workdir/config.sub
@@ -1709,7 +1709,7 @@ case $os in
 	     | bosx* | nextstep* | cxux* | aout* | elf* | oabi* \
 	     | ptx* | coff* | ecoff* | winnt* | domain* | vsta* \
 	     | udi* | lites* | ieee* | go32* | aux* | hcos* \
-	     | chorusrdb* | cegcc* | glidix* \
+	     | chorusrdb* | cegcc* | glidix* | vos* \
 	     | cygwin* | msys* | pe* | moss* | proelf* | rtems* \
 	     | midipix* | mingw32* | mingw64* | mint* \
 	     | uxpv* | beos* | mpeix* | udk* | moxiebox* \
diff --git a/ports/binutils-orig/gas/configure.tgt b/ports/binutils-workdir/gas/configure.tgt
index 6f46e0a..23921f1 100644
--- a/ports/binutils-orig/gas/configure.tgt
+++ b/ports/binutils-workdir/gas/configure.tgt
@@ -221,6 +221,7 @@ case ${generic_target} in
   i386-*-beos*)				fmt=elf ;;
   i386-*-elfiamcu)			fmt=elf arch=iamcu ;;
   i386-*-elf*)				fmt=elf ;;
+  i386-*-vos*)				fmt=elf ;;
   i386-*-fuchsia*)			fmt=elf ;;
   i386-*-bsd*)				fmt=aout em=386bsd ;;
   i386-*-netbsdpe*)			fmt=coff em=pe ;;
diff --git a/ports/binutils-orig/ld/configure.tgt b/ports/binutils-workdir/ld/configure.tgt
index 0c780b2..8407607 100644
--- a/ports/binutils-orig/ld/configure.tgt
+++ b/ports/binutils-workdir/ld/configure.tgt
@@ -319,6 +319,10 @@ i[3-7]86-*-bsd386)	targ_emul=i386bsd
 i[3-7]86-*-bsdi*)	targ_emul=i386bsd
 			targ_extra_ofiles=
 			;;
+x86_64-*-vos*)
+			targ_emul=elf_x86_64
+			targ_extra_emuls=elf_i386
+			;;
 i[3-7]86-*-linux-*)	targ_emul=elf_i386
 			targ_extra_emuls="elf_iamcu"
 			targ64_extra_emuls="elf_x86_64 elf32_x86_64 elf_l1om elf_k1om"
