diff -urN --no-dereference binutils-clean/bfd/config.bfd binutils-workdir/bfd/config.bfd
--- binutils-clean/bfd/config.bfd	2025-02-02 01:00:00.000000000 +0100
+++ binutils-workdir/bfd/config.bfd	2025-04-26 14:41:20.143032608 +0200
@@ -665,6 +665,11 @@
     targ_selvecs=
     targ64_selvecs=x86_64_elf64_vec
     ;;
+  i[3-7]86-*-vinix*)
+    targ_defvec=i386_elf32_vec
+    targ_selvecs=
+    targ64_selvecs=x86_64_elf64_vec
+    ;;
 #ifdef BFD64
   x86_64-*-cloudabi*)
     targ_defvec=x86_64_elf64_cloudabi_vec
@@ -735,6 +740,11 @@
     targ_selvecs="i386_elf32_vec iamcu_elf32_vec x86_64_elf32_vec"
     want64=true
     ;;
+  x86_64-*-vinix*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
 #endif
   i[3-7]86-*-lynxos*)
     targ_defvec=i386_elf32_vec
diff -urN --no-dereference binutils-clean/configure binutils-workdir/configure
--- binutils-clean/configure	2025-02-02 01:00:00.000000000 +0100
+++ binutils-workdir/configure	2025-04-26 15:17:58.611614238 +0200
@@ -3083,7 +3083,7 @@
       *-*-elf* | *-*-sysv4* | *-*-unixware* | *-*-eabi* | hppa*64*-*-hpux* \
       | *-*-linux* | *-*-gnu* | frv-*-uclinux* | *-*-irix5* | *-*-irix6* \
       | *-*-netbsd* | *-*-openbsd* | *-*-freebsd* | *-*-dragonfly* \
-      | *-*-solaris2* | *-*-nto* | *-*-nacl* | *-*-haiku*)
+      | *-*-solaris2* | *-*-nto* | *-*-nacl* | *-*-haiku* | *-*-mlibc*)
         case "${target}" in
           *-*-linux*aout* | *-*-linux*oldld*)
             ;;
diff -urN --no-dereference binutils-clean/gas/configure.tgt binutils-workdir/gas/configure.tgt
--- binutils-clean/gas/configure.tgt	2025-02-02 01:00:00.000000000 +0100
+++ binutils-workdir/gas/configure.tgt	2025-04-26 14:41:20.143369422 +0200
@@ -229,6 +229,7 @@
   i386-*-beos*)				fmt=elf ;;
   i386-*-elfiamcu)			fmt=elf arch=iamcu ;;
   i386-*-elf*)				fmt=elf ;;
+  i386-*-vinix*)				fmt=elf em=gnu ;;
   i386-*-fuchsia*)			fmt=elf ;;
   i386-*-haiku*)			fmt=elf em=haiku ;;
   i386-*-genode*)			fmt=elf ;;
diff -urN --no-dereference binutils-clean/ld/configure.tgt binutils-workdir/ld/configure.tgt
--- binutils-clean/ld/configure.tgt	2025-02-02 01:00:00.000000000 +0100
+++ binutils-workdir/ld/configure.tgt	2025-04-26 14:41:20.143430958 +0200
@@ -384,6 +384,9 @@
 i[3-7]86-*-redox*)	targ_emul=elf_i386
 			targ_extra_emuls=elf_x86_64
 			;;
+i[3-7]86-*-vinix*)	targ_emul=elf_i386
+			targ_extra_emuls=elf_x86_64
+			;;
 i[3-7]86-*-solaris2*)	targ_emul=elf_i386_sol2
 			targ_extra_emuls="elf_i386_ldso elf_i386 elf_iamcu elf_x86_64_sol2 elf_x86_64"
 			targ_extra_libpath=$targ_extra_emuls
@@ -1019,6 +1022,9 @@
 x86_64-*-redox*)	targ_emul=elf_x86_64
 			targ_extra_emuls=elf_i386
 			;;
+x86_64-*-vinix*)	targ_emul=elf_x86_64
+			targ_extra_emuls=elf_i386
+			;;
 x86_64-*-solaris2*)	targ_emul=elf_x86_64_sol2
 			targ_extra_emuls="elf_x86_64 elf_i386_sol2 elf_i386_ldso elf_i386 elf_iamcu"
 			targ_extra_libpath=$targ_extra_emuls
