diff --git mesa-clean/include/drm-uapi/drm.h mesa-workdir/include/drm-uapi/drm.h
index 5e54c3a..085b5b7 100644
--- mesa-clean/include/drm-uapi/drm.h
+++ mesa-workdir/include/drm-uapi/drm.h
@@ -44,7 +44,7 @@ typedef unsigned int drm_handle_t;
 #else /* One of the BSDs */
 
 #include <stdint.h>
-#include <sys/ioccom.h>
+//#include <sys/ioccom.h>
 #include <sys/types.h>
 typedef int8_t   __s8;
 typedef uint8_t  __u8;
diff --git mesa-clean/src/compiler/spirv/spirv_to_nir.c mesa-workdir/src/compiler/spirv/spirv_to_nir.c
index 71cdc83..3d26705 100644
--- mesa-clean/src/compiler/spirv/spirv_to_nir.c
+++ mesa-workdir/src/compiler/spirv/spirv_to_nir.c
@@ -37,6 +37,7 @@
 #include "util/u_string.h"
 
 #include <stdio.h>
+#include <strings.h>
 
 #ifndef NDEBUG
 static enum nir_spirv_debug_level
diff --git mesa-clean/src/egl/main/egllog.c mesa-workdir/src/egl/main/egllog.c
index 984dd5b..6a91952 100644
--- mesa-clean/src/egl/main/egllog.c
+++ mesa-workdir/src/egl/main/egllog.c
@@ -39,6 +39,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include "c11/threads.h"
 #include "util/macros.h"
 #include "util/u_string.h"
diff --git mesa-clean/src/gallium/drivers/llvmpipe/lp_texture.c mesa-workdir/src/gallium/drivers/llvmpipe/lp_texture.c
index 5bfc8db..d2b31e1 100644
--- mesa-clean/src/gallium/drivers/llvmpipe/lp_texture.c
+++ mesa-workdir/src/gallium/drivers/llvmpipe/lp_texture.c
@@ -1102,7 +1102,6 @@ llvmpipe_resource_get_param(struct pipe_screen *screen,
    default:
       break;
    }
-   assert(0);
    *value = 0;
    return false;
 }
diff --git mesa-clean/src/util/debug.c mesa-workdir/src/util/debug.c
index 89ae613..fbf45f4 100644
--- mesa-clean/src/util/debug.c
+++ mesa-workdir/src/util/debug.c
@@ -23,6 +23,7 @@
 
 #include <errno.h>
 #include <string.h>
+#include <strings.h>
 #include "debug.h"
 #include "u_string.h"
 
diff --git mesa-clean/src/util/detect_os.h mesa-workdir/src/util/detect_os.h
index 6506948..a631d69 100644
--- mesa-clean/src/util/detect_os.h
+++ mesa-workdir/src/util/detect_os.h
@@ -81,6 +81,11 @@
 #define DETECT_OS_UNIX 1
 #endif
 
+#if defined(__vinix__)
+#define DETECT_OS_VINIX 1
+#define DETECT_OS_UNIX 1
+#endif
+
 
 /*
  * Make sure DETECT_OS_* are always defined, so that they can be used with #if
@@ -127,5 +132,8 @@
 #ifndef DETECT_OS_WINDOWS
 #define DETECT_OS_WINDOWS 0
 #endif
+#ifndef DETECT_OS_VINIX
+#define DETECT_OS_VINIX 0
+#endif
 
 #endif /* DETECT_OS_H */
diff --git mesa-clean/src/util/os_misc.c mesa-workdir/src/util/os_misc.c
index 31f1c55..0095e53 100644
--- mesa-clean/src/util/os_misc.c
+++ mesa-workdir/src/util/os_misc.c
@@ -57,7 +57,7 @@
 #  include <unistd.h>
 #  include <log/log.h>
 #  include <cutils/properties.h>
-#elif DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD
+#elif DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_VINIX
 #  include <unistd.h>
 #elif DETECT_OS_OPENBSD || DETECT_OS_FREEBSD
 #  include <sys/resource.h>
@@ -223,7 +223,7 @@ os_get_option(const char *name)
 bool
 os_get_total_physical_memory(uint64_t *size)
 {
-#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD
+#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_HURD || DETECT_OS_VINIX
    const long phys_pages = sysconf(_SC_PHYS_PAGES);
    const long page_size = sysconf(_SC_PAGE_SIZE);
 
diff --git mesa-clean/src/util/os_time.c mesa-workdir/src/util/os_time.c
index d2edd88..d7c3dfb 100644
--- mesa-clean/src/util/os_time.c
+++ mesa-workdir/src/util/os_time.c
@@ -53,7 +53,7 @@
 int64_t
 os_time_get_nano(void)
 {
-#if DETECT_OS_LINUX || DETECT_OS_BSD
+#if DETECT_OS_LINUX || DETECT_OS_BSD || DETECT_OS_VINIX
 
    struct timespec tv;
    clock_gettime(CLOCK_MONOTONIC, &tv);
@@ -92,7 +92,7 @@ os_time_get_nano(void)
 void
 os_time_sleep(int64_t usecs)
 {
-#if DETECT_OS_LINUX
+#if DETECT_OS_LINUX || DETECT_OS_VINIX
    struct timespec time;
    time.tv_sec = usecs / 1000000;
    time.tv_nsec = (usecs % 1000000) * 1000;
diff --git mesa-clean/src/util/u_printf.h mesa-workdir/src/util/u_printf.h
index 44dcce5..e9e23ba 100644
--- mesa-clean/src/util/u_printf.h
+++ mesa-workdir/src/util/u_printf.h
@@ -22,6 +22,8 @@
 #ifndef U_PRINTF_H
 #define U_PRINTF_H
 
+#include <stdarg.h>
+
 #ifdef __cplusplus
 
 #include <string>
diff --git mesa-clean/src/util/u_thread.h mesa-workdir/src/util/u_thread.h
index 013e8be..a3a6031 100644
--- mesa-clean/src/util/u_thread.h
+++ mesa-workdir/src/util/u_thread.h
@@ -129,7 +129,7 @@ static inline thrd_t u_thread_create(int (*routine)(void *), void *param)
 static inline void u_thread_setname( const char *name )
 {
 #if defined(HAVE_PTHREAD)
-#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS
+#if DETECT_OS_LINUX || DETECT_OS_CYGWIN || DETECT_OS_SOLARIS || DETECT_OS_VINIX
    int ret = pthread_setname_np(pthread_self(), name);
    if (ret == ERANGE) {
       char buf[16];
@@ -242,7 +242,7 @@ util_set_current_thread_affinity(const uint32_t *mask,
 static inline int64_t
 util_thread_get_time_nano(thrd_t thread)
 {
-#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__)
+#if defined(HAVE_PTHREAD) && !defined(__APPLE__) && !defined(__HAIKU__) && !defined(__vinix__)
    struct timespec ts;
    clockid_t cid;
 
