diff --git a/util/mmap-alloc.c b/util/mmap-alloc.c
index 5b90cb6..ea7a06b 100644
--- a/util/mmap-alloc.c
+++ b/util/mmap-alloc.c
@@ -10,14 +10,16 @@
  * later.  See the COPYING file in the top-level directory.
  */
 
+#include "qemu/osdep.h"
+
 #ifdef CONFIG_LINUX
 #include <linux/mman.h>
+#include <asm-generic/mman.h> /* for ppc64le */
 #else  /* !CONFIG_LINUX */
 #define MAP_SYNC              0x0
 #define MAP_SHARED_VALIDATE   0x0
 #endif /* CONFIG_LINUX */
 
-#include "qemu/osdep.h"
 #include "qemu/mmap-alloc.h"
 #include "qemu/host-utils.h"
 #include "qemu/cutils.h"
