diff -urN --no-dereference libxcb-clean/configure libxcb-workdir/configure
--- libxcb-clean/configure
+++ libxcb-workdir/configure
@@ -21028,7 +21028,7 @@
 NEEDED="xau >= 0.99.2"
 case $host_os in
 linux*|darwin*|solaris*|dragonfly*|freebsd*|netbsd*) ;;
-     *) NEEDED="$NEEDED pthread-stubs" ;;
+     *) NEEDED="$NEEDED" ;;
 esac
 
 pkg_failed=no
diff -urN --no-dereference libxcb-clean/doc/Makefile.in libxcb-workdir/doc/Makefile.in
--- libxcb-clean/doc/Makefile.in
+++ libxcb-workdir/doc/Makefile.in
@@ -518,8 +518,8 @@
 	  $(mkinstalldirs) '$(DESTDIR)$(htmldir)'; \
 	fi
 	@for d in $(docdirs); do \
-	  echo "cp -pR $$d '$(DESTDIR)$(htmldir)/'"; \
-	  cp -pR $$d '$(DESTDIR)$(htmldir)/'; \
+	  echo "cp -R $$d '$(DESTDIR)$(htmldir)/'"; \
+	  cp -R $$d '$(DESTDIR)$(htmldir)/'; \
 	done
 
 uninstall-local:
