diff --git automake-clean/lib/config.guess automake-workdir/lib/config.guess
index e81d3ae..08139ca 100755
--- automake-clean/lib/config.guess
+++ automake-workdir/lib/config.guess
@@ -929,6 +929,9 @@ EOF
     i*:PW*:*)
 	GUESS=$UNAME_MACHINE-pc-pw32
 	;;
+    *:Vinix:*:*)
+	GUESS=$UNAME_MACHINE-pc-vinix
+	;;
     *:Interix*:*)
 	case $UNAME_MACHINE in
 	    x86)
diff --git automake-clean/lib/config.sub automake-workdir/lib/config.sub
index d74fb6d..64292fe 100755
--- automake-clean/lib/config.sub
+++ automake-workdir/lib/config.sub
@@ -1723,7 +1723,7 @@ case $os in
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
-	     | os9* | macos* | osx* | ios* \
+	     | os9* | macos* | osx* | ios* | vinix* \
 	     | mpw* | magic* | mmixware* | mon960* | lnews* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* | twizzler* \
