diff -urN --no-dereference libzip-clean/lib/zip_close.c libzip-workdir/lib/zip_close.c
--- libzip-clean/lib/zip_close.c
+++ libzip-workdir/lib/zip_close.c
@@ -36,6 +36,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <strings.h>
 #ifdef _WIN32
 #include <fcntl.h>
 #include <io.h>
diff -urN --no-dereference libzip-clean/lib/zip_name_locate.c libzip-workdir/lib/zip_name_locate.c
--- libzip-clean/lib/zip_name_locate.c
+++ libzip-workdir/lib/zip_name_locate.c
@@ -33,9 +33,9 @@
 
 
 #include <string.h>
-#ifdef HAVE_STRINGS_H
+//#ifdef HAVE_STRINGS_H
 #include <strings.h>
-#endif
+//#endif
 
 #include "zipint.h"
 
diff -urN --no-dereference libzip-clean/src/ziptool.c libzip-workdir/src/ziptool.c
--- libzip-clean/src/ziptool.c
+++ libzip-workdir/src/ziptool.c
@@ -38,6 +38,7 @@
 #include <errno.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <time.h>
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
