diff --git libzip-clean/lib/zip_close.c libzip-workdir/lib/zip_close.c
index ddc2c24..6206bd0 100644
--- libzip-clean/lib/zip_close.c
+++ libzip-workdir/lib/zip_close.c
@@ -36,6 +36,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <strings.h>
 #ifdef _WIN32
 #include <fcntl.h>
 #include <io.h>
@@ -742,4 +743,4 @@ static int torrentzip_compare_names(const void *a, const void *b) {
     }
 
     return strcasecmp(aname, bname);
-}
\ No newline at end of file
+}
diff --git libzip-clean/lib/zip_name_locate.c libzip-workdir/lib/zip_name_locate.c
index 4248dc2..fdf113a 100644
--- libzip-clean/lib/zip_name_locate.c
+++ libzip-workdir/lib/zip_name_locate.c
@@ -33,9 +33,9 @@
 
 
 #include <string.h>
-#ifdef HAVE_STRINGS_H
+//#ifdef HAVE_STRINGS_H
 #include <strings.h>
-#endif
+//#endif
 
 #include "zipint.h"
 
diff --git libzip-clean/src/ziptool.c libzip-workdir/src/ziptool.c
index 281bfbb..9897044 100644
--- libzip-clean/src/ziptool.c
+++ libzip-workdir/src/ziptool.c
@@ -38,6 +38,7 @@
 #include <errno.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <time.h>
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>
