diff -urN --no-dereference glib-clean/gio/gcredentialsprivate.h glib-workdir/gio/gcredentialsprivate.h
--- glib-clean/gio/gcredentialsprivate.h
+++ glib-workdir/gio/gcredentialsprivate.h
@@ -104,7 +104,7 @@
  */
 #undef G_CREDENTIALS_HAS_PID
 
-#ifdef __linux__
+#if defined(__linux__) || defined(__vinix__)
 #define G_CREDENTIALS_SUPPORTED 1
 #define G_CREDENTIALS_USE_LINUX_UCRED 1
 #define G_CREDENTIALS_NATIVE_TYPE G_CREDENTIALS_TYPE_LINUX_UCRED
diff -urN --no-dereference glib-clean/glib/glib-init.c glib-workdir/glib/glib-init.c
--- glib-clean/glib/glib-init.c
+++ glib-workdir/glib/glib-init.c
@@ -31,6 +31,7 @@
 #include "gmem.h"       /* for g_mem_gc_friendly */
 
 #include <string.h>
+#include <strings.h>
 #include <stdlib.h>
 #include <stdio.h>
 #include <ctype.h>
diff -urN --no-dereference glib-clean/glib/gstrfuncs.c glib-workdir/glib/gstrfuncs.c
--- glib-clean/glib/gstrfuncs.c
+++ glib-workdir/glib/gstrfuncs.c
@@ -35,6 +35,7 @@
 #include <stdlib.h>
 #include <locale.h>
 #include <string.h>
+#include <strings.h>
 #include <locale.h>
 #include <errno.h>
 #include <garray.h>
