diff --git a/3rdparty/bash-orig/builtins/psize.c b/3rdparty/bash-workdir/builtins/psize.c
index 30881fb..be9c84b 100644
--- a/3rdparty/bash-orig/builtins/psize.c
+++ b/3rdparty/bash-workdir/builtins/psize.c
@@ -21,25 +21,12 @@
 /*  Write output in 128-byte chunks until we get a sigpipe or write gets an
     EPIPE.  Then report how many bytes we wrote.  We assume that this is the
     pipe size. */
-#include <config.h>
-
-#if defined (HAVE_UNISTD_H)
-#  ifdef _MINIX
-#    include <sys/types.h>
-#  endif
-#  include <unistd.h>
-#endif
 
 #include <stdio.h>
-#ifndef _MINIX
-#include "../bashtypes.h"
-#endif
+#include <stdlib.h>
 #include <signal.h>
 #include <errno.h>
-
-#include "../command.h"
-#include "../general.h"
-#include "../sig.h"
+#include <unistd.h>
 
 #ifndef errno
 extern int errno;
@@ -47,7 +34,7 @@ extern int errno;
 
 int nw;
 
-sighandler
+void
 sigpipe (sig)
      int sig;
 {
diff --git a/3rdparty/bash-orig/lib/readline/terminal.c b/3rdparty/bash-workdir/lib/readline/terminal.c
index 05415dc..2f0c770 100644
--- a/3rdparty/bash-orig/lib/readline/terminal.c
+++ b/3rdparty/bash-workdir/lib/readline/terminal.c
@@ -103,7 +103,7 @@ static char *term_string_buffer = (char *)NULL;
 static int tcap_initialized;
 
 #if !defined (__linux__) && !defined (NCURSES_VERSION)
-#  if defined (__EMX__) || defined (NEED_EXTERN_PC)
+#  if defined (__EMX__) || defined (__vinix__) || defined (NEED_EXTERN_PC)
 extern 
 #  endif /* __EMX__ || NEED_EXTERN_PC */
 char PC, *BC, *UP;
diff --git a/3rdparty/bash-orig/lib/termcap/termcap.c b/3rdparty/bash-workdir/lib/termcap/termcap.c
index ba3dab2..2882f0c 100644
--- a/3rdparty/bash-orig/lib/termcap/termcap.c
+++ b/3rdparty/bash-workdir/lib/termcap/termcap.c
@@ -627,7 +627,7 @@ scan_file (str, fd, bufp)
   bufp->ateof = 0;
   *bufp->ptr = '\0';
 
-  lseek (fd, 0L, 0);
+  lseek (fd, 0L, SEEK_SET);
 
   while (!bufp->ateof)
     {
diff --git a/3rdparty/bash-orig/mksyntax.c b/3rdparty/bash-workdir/mksyntax.c
index 0385686..1a73eca 100644
--- a/3rdparty/bash-orig/mksyntax.c
+++ b/3rdparty/bash-workdir/mksyntax.c
@@ -20,16 +20,12 @@
    along with Bash.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#include "config.h"
-
 #include <stdio.h>
+#include <string.h>
 #include "bashansi.h"
 #include "chartypes.h"
 #include <errno.h>
-
-#ifdef HAVE_UNISTD_H
-#  include <unistd.h>
-#endif
+#include <unistd.h>
 
 #include "syntax.h"
 
@@ -40,9 +36,7 @@ extern char *optarg;
 extern int errno;
 #endif
 
-#ifndef HAVE_STRERROR
 extern char *strerror();
-#endif
 
 struct wordflag {
 	int	flag;
@@ -375,41 +369,3 @@ main(argc, argv)
     fclose (fp);
   exit (0);
 }
-
-
-#if !defined (HAVE_STRERROR)
-
-#include <bashtypes.h>
-#if defined (HAVE_SYS_PARAM_H)
-#  include <sys/param.h>
-#endif
-
-#if defined (HAVE_UNISTD_H)
-#  include <unistd.h>
-#endif
-
-/* Return a string corresponding to the error number E.  From
-   the ANSI C spec. */
-#if defined (strerror)
-#  undef strerror
-#endif
-
-char *
-strerror (e)
-     int e;
-{
-  static char emsg[40];
-#if defined (HAVE_SYS_ERRLIST)
-  extern int sys_nerr;
-  extern char *sys_errlist[];
-
-  if (e > 0 && e < sys_nerr)
-    return (sys_errlist[e]);
-  else
-#endif /* HAVE_SYS_ERRLIST */
-    {
-      sprintf (emsg, "Unknown system error %d", e);
-      return (&emsg[0]);
-    }
-}
-#endif /* HAVE_STRERROR */
diff --git a/3rdparty/bash-orig/shell.c b/3rdparty/bash-workdir/shell.c
index ce8087f..b7475dd 100644
--- a/3rdparty/bash-orig/shell.c
+++ b/3rdparty/bash-workdir/shell.c
@@ -1614,7 +1614,7 @@ open_shell_script (script_name)
 #endif
 
   /* Only do this with non-tty file descriptors we can seek on. */
-  if (fd_is_tty == 0 && (lseek (fd, 0L, 1) != -1))
+  if (fd_is_tty == 0 && (lseek (fd, 0L, SEEK_CUR) != -1))
     {
       /* Check to see if the `file' in `bash file' is a binary file
 	 according to the same tests done by execute_simple_command (),
@@ -1651,7 +1651,7 @@ open_shell_script (script_name)
 	  exit (EX_BINARY_FILE);
 	}
       /* Now rewind the file back to the beginning. */
-      lseek (fd, 0L, 0);
+      lseek (fd, 0L, SEEK_SET);
     }
 
   /* Open the script.  But try to move the file descriptor to a randomly
diff --git a/3rdparty/bash-orig/support/bashversion.c b/3rdparty/bash-workdir/support/bashversion.c
index 4f86b13..64779de 100644
--- a/3rdparty/bash-orig/support/bashversion.c
+++ b/3rdparty/bash-workdir/support/bashversion.c
@@ -18,15 +18,12 @@
    along with Bash.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#include "config.h"
-
 #include "stdc.h"
 
 #include <stdio.h>
+#include <string.h>
 
-#if defined (HAVE_UNISTD_H)
-#  include <unistd.h>
-#endif
+#include <unistd.h>
 
 #include "bashansi.h"
 
diff --git a/3rdparty/bash-orig/support/config.sub b/3rdparty/bash-workdir/support/config.sub
index c874b7a..b6f5162 100755
--- a/3rdparty/bash-orig/support/config.sub
+++ b/3rdparty/bash-workdir/support/config.sub
@@ -1707,7 +1707,7 @@ case $os in
 	     | bosx* | nextstep* | cxux* | aout* | elf* | oabi* \
 	     | ptx* | coff* | ecoff* | winnt* | domain* | vsta* \
 	     | udi* | lites* | ieee* | go32* | aux* | hcos* \
-	     | chorusrdb* | cegcc* | glidix* \
+	     | chorusrdb* | cegcc* | glidix* | vinix* \
 	     | cygwin* | msys* | pe* | moss* | proelf* | rtems* \
 	     | midipix* | mingw32* | mingw64* | mint* \
 	     | uxpv* | beos* | mpeix* | udk* | moxiebox* \
diff --git a/3rdparty/bash-orig/support/man2html.c b/3rdparty/bash-workdir/support/man2html.c
index e6f441b..f02453d 100644
--- a/3rdparty/bash-orig/support/man2html.c
+++ b/3rdparty/bash-workdir/support/man2html.c
@@ -63,7 +63,7 @@
  *    time to look through all the available manpages.)
  */
 #ifdef HAVE_CONFIG_H
-#include <config.h>
+//#include <config.h>
 #endif
 
 #define NROFF 0
@@ -123,7 +123,8 @@ char *manpage;
 #define BD_LITERAL  1
 #define BD_INDENT   2
 
-#ifndef HAVE_STRERROR
+//#ifndef HAVE_STRERROR
+#if (0)
 static char *
 strerror(int e)
 {
diff --git a/3rdparty/bash-orig/support/mksignames.c b/3rdparty/bash-workdir/support/mksignames.c
index ba87ae8..4b39787 100644
--- a/3rdparty/bash-orig/support/mksignames.c
+++ b/3rdparty/bash-workdir/support/mksignames.c
@@ -19,8 +19,6 @@
    along with Bash.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#include <config.h>
-
 #include <sys/types.h>
 #include <signal.h>
 
diff --git a/3rdparty/bash-orig/support/signames.c b/3rdparty/bash-workdir/support/signames.c
index aba4842..b4bbdc6 100644
--- a/3rdparty/bash-orig/support/signames.c
+++ b/3rdparty/bash-workdir/support/signames.c
@@ -18,8 +18,6 @@
    along with Bash.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#include <config.h>
-
 #include <stdio.h>
 
 #include <sys/types.h>
