From 06d0415700517dd307e133db4a12d838b3c34cf7 Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Thu, 11 Nov 2021 00:51:34 +0100
Subject: [PATCH] Vinix specific changes

---
 Makefile.am | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.am b/Makefile.am
index 8e96ebe..3df5468 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -19,7 +19,7 @@
 ALL_RECURSIVE_TARGETS =
 
 EXTRA_DIST = bootstrap exgettext ChangeLog-2008 cfg.mk dist-check.mk
-SUBDIRS = lib src tests doc man po gnulib-tests
+SUBDIRS = lib src tests doc po gnulib-tests
 
 ACLOCAL_AMFLAGS = -I m4
 AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS)
-- 
2.33.1

