From a8bc1170e15f54c6b76e658d582065d74222edf2 Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Thu, 9 Sep 2021 06:18:11 +0200
Subject: [PATCH] Vinix specific changes

---
 Makefile.am | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.am b/Makefile.am
index 5041753..5da919b 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -19,7 +19,7 @@
 ALL_RECURSIVE_TARGETS =
 
 EXTRA_DIST = bootstrap exgettext ChangeLog-2008 cfg.mk dist-check.mk
-SUBDIRS = lib src tests doc man po gnulib-tests
+SUBDIRS = lib src tests doc po gnulib-tests
 
 ACLOCAL_AMFLAGS = -I m4
 AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS)
-- 
2.33.0

