diff --git a/3rdparty/diffutils-orig/build-aux/config.sub b/3rdparty/diffutils-workdir/build-aux/config.sub
index a5081b8..f15f483 100755
--- a/3rdparty/diffutils-orig/build-aux/config.sub
+++ b/3rdparty/diffutils-workdir/build-aux/config.sub
@@ -1361,7 +1361,7 @@ case $os in
 	     | powermax* | dnix* | nx6 | nx7 | sei* | dragonfly* \
 	     | skyos* | haiku* | rdos* | toppers* | drops* | es* \
 	     | onefs* | tirtos* | phoenix* | fuchsia* | redox* | bme* \
-	     | midnightbsd* | amdhsa* | unleashed*)
+	     | midnightbsd* | amdhsa* | unleashed* | vinix*)
 	# Remember, each alternative MUST END IN *, to match a version number.
 		;;
 	qnx*)
