diff -urN --no-dereference dbus-clean/dbus/dbus-sysdeps-unix.c dbus-workdir/dbus/dbus-sysdeps-unix.c
--- dbus-clean/dbus/dbus-sysdeps-unix.c
+++ dbus-workdir/dbus/dbus-sysdeps-unix.c
@@ -5024,7 +5024,7 @@
 
   if (_DBUS_UNLIKELY (!check_setuid_initialised))
     {
-#ifdef HAVE_GETRESUID
+#if defined HAVE_GETRESUID && !defined(__vinix__)
       if (getresuid (&ruid, &euid, &suid) != 0 ||
           getresgid (&rgid, &egid, &sgid) != 0)
 #endif /* HAVE_GETRESUID */
