diff -urN --no-dereference coreutils-clean/gnulib-tests/getlocalename_l-unsafe.c coreutils-workdir/gnulib-tests/getlocalename_l-unsafe.c
--- coreutils-clean/gnulib-tests/getlocalename_l-unsafe.c	2025-04-06 00:23:58.000000000 +0200
+++ coreutils-workdir/gnulib-tests/getlocalename_l-unsafe.c	2025-04-26 19:02:23.700106950 +0200
@@ -469,6 +469,8 @@
       struct gl_locale_category_t *plc =
         &locale->category[gl_log2_lcmask_to_index (gl_log2_lc_mask (category))];
       return (struct string_with_storage) { plc->name, STORAGE_OBJECT };
+#elif defined __vinix__
+      return (struct string_with_storage) { "C", STORAGE_INDEFINITE };
 #elif __GLIBC__ >= 2 && !defined __UCLIBC__
       /* Work around an incorrect definition of the _NL_LOCALE_NAME macro in
          glibc < 2.12.
diff -urN --no-dereference coreutils-clean/lib/getlocalename_l-unsafe.c coreutils-workdir/lib/getlocalename_l-unsafe.c
--- coreutils-clean/lib/getlocalename_l-unsafe.c	2025-04-06 00:23:58.000000000 +0200
+++ coreutils-workdir/lib/getlocalename_l-unsafe.c	2025-04-26 19:00:42.421261271 +0200
@@ -469,6 +469,8 @@
       struct gl_locale_category_t *plc =
         &locale->category[gl_log2_lcmask_to_index (gl_log2_lc_mask (category))];
       return (struct string_with_storage) { plc->name, STORAGE_OBJECT };
+#elif defined __vinix__
+      return (struct string_with_storage) { "C", STORAGE_INDEFINITE };
 #elif __GLIBC__ >= 2 && !defined __UCLIBC__
       /* Work around an incorrect definition of the _NL_LOCALE_NAME macro in
          glibc < 2.12.
