diff -urN --no-dereference xf86-video-fbdev-clean/src/Makefile.in xf86-video-fbdev-workdir/src/Makefile.in
--- xf86-video-fbdev-clean/src/Makefile.in
+++ xf86-video-fbdev-workdir/src/Makefile.in
@@ -373,7 +373,7 @@
 # TODO: -nostdlib/-Bstatic/-lgcc platform magic, not installing the .a, etc.
 AM_CFLAGS = $(BASE_CFLAGS) $(XORG_CFLAGS)
 fbdev_drv_la_LTLIBRARIES = fbdev_drv.la
-fbdev_drv_la_LDFLAGS = -module -avoid-version
+fbdev_drv_la_LDFLAGS = -module -avoid-version -R@moduledir@ -L${SYSROOT}@moduledir@ -lfbdevhw -lshadow
 fbdev_drv_ladir = @moduledir@/drivers
 fbdev_drv_la_SOURCES = \
          fbdev.c
diff -urN --no-dereference xf86-video-fbdev-clean/src/fbdev.c xf86-video-fbdev-workdir/src/fbdev.c
--- xf86-video-fbdev-clean/src/fbdev.c
+++ xf86-video-fbdev-workdir/src/fbdev.c
@@ -331,7 +331,7 @@
 
 	    dev = xf86FindOptionValue(devSections[i]->options,"fbdev");
 	    if (devSections[i]->busID) {
-#ifndef XSERVER_LIBPCIACCESS
+#ifdef XSERVER_LIBPCIACCESS
 	        if (xf86ParsePciBusString(devSections[i]->busID,&bus,&device,
 					  &func)) {
 		    if (!xf86CheckPciSlot(bus,device,func))
@@ -343,7 +343,7 @@
 	    if (fbdevHWProbe(NULL,dev,NULL)) {
 		pScrn = NULL;
 		if (isPci) {
-#ifndef XSERVER_LIBPCIACCESS
+#ifdef XSERVER_LIBPCIACCESS
 		    /* XXX what about when there's no busID set? */
 		    int entity;
 		    
@@ -419,6 +419,7 @@
 	fPtr->pEnt = xf86GetEntityInfo(pScrn->entityList[0]);
 
 #ifndef XSERVER_LIBPCIACCESS
+#if 0
 	pScrn->racMemFlags = RAC_FB | RAC_COLORMAP | RAC_CURSOR | RAC_VIEWPORT;
 	/* XXX Is this right?  Can probably remove RAC_FB */
 	pScrn->racIoFlags = RAC_FB | RAC_COLORMAP | RAC_CURSOR | RAC_VIEWPORT;
@@ -429,6 +430,7 @@
 		   "xf86RegisterResources() found resource conflicts\n");
 		return FALSE;
 	}
+#endif
 #else
 	if (fPtr->pEnt->location.type == BUS_PCI)
 	    pci_dev = fPtr->pEnt->location.id.pci;
