diff --git libtool-clean/libtoolize.in libtool-workdir/libtoolize.in
index 0c40fed..763619b 100644
--- libtool-clean/libtoolize.in
+++ libtool-workdir/libtoolize.in
@@ -1891,7 +1891,7 @@ func_require_seen_libtool ()
   # Do not remove config.guess, config.sub or install-sh, we don't
   # install them without --install, and the project may not be using
   # Automake.  Similarly, do not remove Gnulib files.
-  all_pkgaux_files="compile depcomp missing ltmain.sh"
+  all_pkgaux_files=""
   all_pkgmacro_files="libtool.m4 ltargz.m4 ltdl.m4 ltoptions.m4 ltsugar.m4 ltversion.in ltversion.m4 lt~obsolete.m4"
   all_pkgltdl_files="COPYING.LIB Makefile Makefile.in Makefile.inc Makefile.am README acinclude.m4 aclocal.m4 argz_.h argz.c config.h.in config-h.in configure configure.ac configure.in libltdl/lt__alloc.h libltdl/lt__argz.h libltdl/lt__dirent.h libltdl/lt__glibc.h libltdl/lt__private.h libltdl/lt__strl.h libltdl/lt_dlloader.h libltdl/lt_error.h libltdl/lt_system.h libltdl/slist.h loaders/dld_link.c loaders/dlopen.c loaders/dyld.c loaders/load_add_on.c loaders/loadlibrary.c loaders/preopen.c loaders/shl_load.c lt__alloc.c lt__argz.c lt__dirent.c lt__strl.c lt_dlloader.c lt_error.c ltdl.c ltdl.h ltdl.mk slist.c"
 
diff --git libtool-clean/m4/libtool.m4 libtool-workdir/m4/libtool.m4
index 79a2451..db95342 100644
--- libtool-clean/m4/libtool.m4
+++ libtool-workdir/m4/libtool.m4
@@ -1696,7 +1696,7 @@ AC_CACHE_VAL([lt_cv_sys_max_cmd_len], [dnl
     lt_cv_sys_max_cmd_len=12288;    # 12K is about right
     ;;
 
-  gnu* | ironclad*)
+  gnu* | ironclad* | vinix*)
     # Under GNU Hurd and Ironclad, this test is not required because there
     # is no limit to the length of command line arguments.
     # Libtool will interpret -1 as no limit whatsoever
