diff --git libtool-clean/m4/libtool.m4 libtool-workdir/m4/libtool.m4
index 79a2451..db95342 100644
--- libtool-clean/m4/libtool.m4
+++ libtool-workdir/m4/libtool.m4
@@ -1696,7 +1696,7 @@ AC_CACHE_VAL([lt_cv_sys_max_cmd_len], [dnl
     lt_cv_sys_max_cmd_len=12288;    # 12K is about right
     ;;
 
-  gnu* | ironclad*)
+  gnu* | ironclad* | vinix*)
     # Under GNU Hurd and Ironclad, this test is not required because there
     # is no limit to the length of command line arguments.
     # Libtool will interpret -1 as no limit whatsoever
