diff --git autoconf-clean/build-aux/config.guess autoconf-workdir/build-aux/config.guess
index cdfc439..02e3643 100755
--- autoconf-clean/build-aux/config.guess
+++ autoconf-workdir/build-aux/config.guess
@@ -4,7 +4,8 @@
 
 # shellcheck disable=SC2006,SC2268 # see below for rationale
 
-timestamp='2023-08-22'
+# timestamp it to always be newer
+timestamp='9999-99-99'
 
 # This file is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License as published by
@@ -976,6 +977,12 @@ EOF
     *:[Mm]anagarm:*:*)
 	GUESS="$UNAME_MACHINE-unknown-managarm-mlibc"
 	;;
+    x86_64:[Vv]inix:*:*|i?86:[Vv]inix:*:*)
+	GUESS="$UNAME_MACHINE-pc-vinix-mlibc"
+	;;
+    *:[Vv]inix:*:*)
+	GUESS="$UNAME_MACHINE-unknown-vinix-mlibc"
+	;;
     *:Minix:*:*)
 	GUESS=$UNAME_MACHINE-unknown-minix
 	;;
diff --git autoconf-clean/build-aux/config.sub autoconf-workdir/build-aux/config.sub
index defe52c..426e8de 100755
--- autoconf-clean/build-aux/config.sub
+++ autoconf-workdir/build-aux/config.sub
@@ -4,7 +4,8 @@
 
 # shellcheck disable=SC2006,SC2268 # see below for rationale
 
-timestamp='2023-09-19'
+# timestamp it to always be newer
+timestamp='9999-99-99'
 
 # This file is free software; you can redistribute it and/or modify it
 # under the terms of the GNU General Public License as published by
@@ -145,7 +146,7 @@ case $1 in
 			nto-qnx* | linux-* | uclinux-uclibc* \
 			| uclinux-gnu* | kfreebsd*-gnu* | knetbsd*-gnu* | netbsd*-gnu* \
 			| netbsd*-eabi* | kopensolaris*-gnu* | cloudabi*-eabi* \
-			| storm-chaos* | os2-emx* | rtmk-nova* | managarm-* \
+			| storm-chaos* | os2-emx* | rtmk-nova* | managarm-* | vinix-* \
 			| windows-* )
 				basic_machine=$field1
 				basic_os=$maybe_os
@@ -1325,6 +1326,10 @@ EOF
 		kernel=managarm
 		os=`echo "$basic_os" | sed -e 's|managarm|mlibc|'`
 		;;
+	vinix*)
+		kernel=vinix
+		os=`echo "$basic_os" | sed -e 's|vinix|mlibc|'`
+		;;
 	*)
 		kernel=
 		os=$basic_os
@@ -1825,6 +1830,8 @@ case $kernel-$os-$obj in
 		;;
 	managarm-mlibc*- | managarm-kernel*- )
 		;;
+	vinix-mlibc*- )
+		;;
 	windows*-msvc*-)
 		;;
 	-dietlibc*- | -newlib*- | -musl*- | -relibc*- | -uclibc*- | -mlibc*- )
