diff -urN --no-dereference ace-of-penguins-clean/lib/Makefile.in ace-of-penguins-workdir/lib/Makefile.in
--- ace-of-penguins-clean/lib/Makefile.in
+++ ace-of-penguins-workdir/lib/Makefile.in
@@ -565,10 +565,10 @@
 	./make-imglib -n cards -i $(srcdir)/png -d images.d $(srcdir)/images.in images.c
 
 make-imglib : make-imglib.c
-	$(BUILD_CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) $(srcdir)/make-imglib.c -o make-imglib -lpng -lz
+	$(BUILD_CC) $(srcdir)/make-imglib.c -o make-imglib -lpng -lz
 
 text2c : text2c.c
-	$(BUILD_CC) $(CFLAGS) $(CPPFLAGS) $(LDFLAGS) $(srcdir)/text2c.c -o text2c -lpng -lz
+	$(BUILD_CC) $(srcdir)/text2c.c -o text2c -lpng -lz
 
 # Tell versions [3.59,3.63) of GNU make to not export all variables.
 # Otherwise a system limit (for SysV at least) may be exceeded.
diff -urN --no-dereference ace-of-penguins-clean/lib/imagelib.c ace-of-penguins-workdir/lib/imagelib.c
--- ace-of-penguins-clean/lib/imagelib.c
+++ ace-of-penguins-workdir/lib/imagelib.c
@@ -1,5 +1,6 @@
 #include <stdio.h>
 #include <string.h>
+#include <stdlib.h>
 #include "imagelib.h"
 #include "cards.h"
 
diff -urN --no-dereference ace-of-penguins-clean/lib/make-imglib.c ace-of-penguins-workdir/lib/make-imglib.c
--- ace-of-penguins-clean/lib/make-imglib.c
+++ ace-of-penguins-workdir/lib/make-imglib.c
@@ -1,6 +1,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <ctype.h>
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <unistd.h>
@@ -86,7 +87,7 @@
     png_ptr = png_create_read_struct (PNG_LIBPNG_VER_STRING, 0, 0, 0);
     info_ptr = png_create_info_struct (png_ptr);
 
-    if (setjmp (png_ptr->jmpbuf)) {
+    if (setjmp (png_jmpbuf(png_ptr))) {
       fclose (f);
       continue;
     }
@@ -202,7 +203,7 @@
   char *rv;
   if (string) {
     next = string;
-    return;
+    return NULL;
   }
   while (*next && !isgraph(*next)) next++;
   if (!*next) return 0;
diff -urN --no-dereference ace-of-penguins-clean/lib/xwin.c ace-of-penguins-workdir/lib/xwin.c
--- ace-of-penguins-clean/lib/xwin.c
+++ ace-of-penguins-workdir/lib/xwin.c
@@ -55,7 +55,7 @@
   { "-visual", OPTION_INTEGER, &visual_id },
   { 0, 0, 0 }
 };
-OptionDesc *xwin_options = xwin_options_list;
+//OptionDesc *xwin_options = xwin_options_list;
 
 Display *display=0;
 int screen=0;
@@ -89,10 +89,10 @@
 /* Motif window hints */
 typedef struct
 {
-  unsigned flags;
-  unsigned functions;
-  unsigned decorations;
-  int inputMode;
+  unsigned long flags;
+  unsigned long functions;
+  unsigned long decorations;
+  long inputMode;
 } PropMotifWmHints;
 
 typedef PropMotifWmHints        PropMwmHints;
@@ -841,13 +841,13 @@
   png_ptr = png_create_read_struct (PNG_LIBPNG_VER_STRING, 0, 0, 0);
   info_ptr = png_create_info_struct (png_ptr);
 
-  if (setjmp (png_ptr->jmpbuf)) {
+  if (setjmp (png_jmpbuf(png_ptr))) {
     fprintf(stderr, "Invalid PNG image!\n");
     return;
   }
 
   file_bytes = src->file_data;
-  png_set_read_fn (png_ptr, (voidp)&file_bytes, (png_rw_ptr)png_reader);
+  png_set_read_fn (png_ptr, (void*)&file_bytes, (png_rw_ptr)png_reader);
 
   png_read_info (png_ptr, info_ptr);
 
