From 44f850e736e8f4ea586a0fda7127f79565488093 Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Sat, 4 Sep 2021 02:31:43 +0200
Subject: [PATCH] Vinix specific changes

---
 lib/config.sub | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/config.sub b/lib/config.sub
index 40ea5df..f75e595 100755
--- a/lib/config.sub
+++ b/lib/config.sub
@@ -1394,7 +1394,7 @@ case $os in
 	-gnu* | -bsd* | -mach* | -minix* | -genix* | -ultrix* | -irix* \
 	      | -*vms* | -sco* | -esix* | -isc* | -aix* | -cnk* | -sunos | -sunos[34]*\
 	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
-	      | -sym* | -kopensolaris* | -plan9* \
+	      | -sym* | -kopensolaris* | -plan9* | -vinix* \
 	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
 	      | -aos* | -aros* | -cloudabi* | -sortix* \
 	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
-- 
2.33.0

