--- a/vlib/v/pref/default.v
+++ b/vlib/v/pref/default.v
@@ -165,8 +165,9 @@ pub fn (mut p Preferences) fill_with_defaults() {
 		p.parse_define('cross') // TODO: remove when `$if cross {` works
 	}
 	if p.gc_mode == .unknown {
-		if p.backend != .c || p.building_v || p.is_bare || p.ccompiler == 'msvc' {
+		if p.backend != .c || p.building_v || p.is_bare || p.ccompiler == 'msvc'
+			|| p.os == .vinix {
 			p.gc_mode = .no_gc
 			p.build_options << ['-gc', 'none']
 		} else {
 			// enable the GC by default
@@ -178,6 +179,10 @@ pub fn (mut p Preferences) fill_with_defaults() {
 			p.parse_define('gcboehm_opt')
 		}
 	}
+	// Vinix does not provide the libc backtrace interfaces yet.
+	if p.os == .vinix {
+		p.parse_define('no_backtrace')
+	}
 	if p.is_debug {
 		p.parse_define('debug')
 	}
