diff --git a/dlls/ntdll/unix/signal_x86_64.c b/dlls/ntdll/unix/signal_x86_64.c
index 2ee9b43b6c2..84f1bea1edc 100644
--- a/dlls/ntdll/unix/signal_x86_64.c
+++ b/dlls/ntdll/unix/signal_x86_64.c
@@ -83,8 +83,9 @@ WINE_DECLARE_DEBUG_CHANNEL(seh);
 /***********************************************************************
  * signal context platform-specific definitions
  */
-#ifdef linux
+#if defined(linux) || defined(__vinix__)
 
+#ifdef linux
 #include <asm/prctl.h>
 static inline int arch_prctl( int func, void *ptr ) { return syscall( __NR_arch_prctl, func, ptr ); }
 
@@ -111,6 +112,7 @@ __ASM_GLOBAL_FUNC( alloc_fs_sel,
                    "popq %r12\n\t"
                    "popq %rbx\n\t"
                    "ret" );
+#endif
 
 #ifndef FP_XSTATE_MAGIC1
 #define FP_XSTATE_MAGIC1 0x46505853
@@ -2588,7 +2590,12 @@ void call_init_thunk( LPTHREAD_START_ROUTINE entry, void *arg, BOOL suspend, TEB
     assert( thread_data->xstate_features_size == xstate_features_size );
     thread_data->instrumentation_callback = &instrumentation_callback;
 
-#if defined __linux__
+#if defined __vinix__
+    register unsigned long syscall_number __asm__("rdi") = 8;
+    register void *syscall_argument __asm__("rsi") = teb;
+    __asm__ volatile ("syscall" : "+D" (syscall_number), "+S" (syscall_argument)
+                               : : "rax", "rcx", "rdx", "r11", "memory");
+#elif defined __linux__
     arch_prctl( ARCH_SET_GS, teb );
     arch_prctl( ARCH_GET_FS, &thread_data->pthread_teb );
     if (fs32_sel) alloc_fs_sel( fs32_sel >> 3, get_wow_teb( teb ));
