From ef06c8ab716d8b6f0b5e7862c4922edcf46909c0 Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Sun, 5 Sep 2021 15:34:54 +0200
Subject: [PATCH] Vinix specific changes

---
 Makefile.devel            | 8 ++++----
 libcharset/Makefile.devel | 6 +++---
 2 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/Makefile.devel b/Makefile.devel
index 2a5e101..a179aaf 100644
--- a/Makefile.devel
+++ b/Makefile.devel
@@ -4,10 +4,10 @@
 
 SHELL = /bin/sh
 MAKE = make
-AUTOCONF = autoconf-2.69
-AUTOHEADER = autoheader-2.69
-AUTOMAKE = automake-1.16
-ACLOCAL = aclocal-1.16
+AUTOCONF = autoconf
+AUTOHEADER = autoheader
+AUTOMAKE = automake
+ACLOCAL = aclocal
 GPERF = gperf
 CC = gcc -Wall
 CFLAGS = -O
diff --git a/libcharset/Makefile.devel b/libcharset/Makefile.devel
index 04f4c7a..3e0e2ca 100644
--- a/libcharset/Makefile.devel
+++ b/libcharset/Makefile.devel
@@ -3,9 +3,9 @@
 
 SHELL = /bin/sh
 MAKE = make
-AUTOCONF = autoconf-2.69
-AUTOHEADER = autoheader-2.69
-ACLOCAL = aclocal-1.16
+AUTOCONF = autoconf
+AUTOHEADER = autoheader
+ACLOCAL = aclocal
 CP = cp
 RM = rm -f
 
-- 
2.33.0

