diff --git a/3rdparty/libiconv-orig/build-aux/config.sub b/3rdparty/libiconv-workdir/build-aux/config.sub
index a44fd8a..b613f42 100755
--- a/3rdparty/libiconv-orig/build-aux/config.sub
+++ b/3rdparty/libiconv-workdir/build-aux/config.sub
@@ -1343,7 +1343,7 @@ case $os in
 	gnu* | bsd* | mach* | minix* | genix* | ultrix* | irix* \
 	     | *vms* | esix* | aix* | cnk* | sunos | sunos[34]*\
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
-	     | sym* | kopensolaris* | plan9* \
+	     | sym* | kopensolaris* | plan9* | vinix* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* \
 	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
diff --git a/3rdparty/libiconv-orig/libcharset/build-aux/config.sub b/3rdparty/libiconv-workdir/libcharset/build-aux/config.sub
index a44fd8a..b613f42 100755
--- a/3rdparty/libiconv-orig/libcharset/build-aux/config.sub
+++ b/3rdparty/libiconv-workdir/libcharset/build-aux/config.sub
@@ -1343,7 +1343,7 @@ case $os in
 	gnu* | bsd* | mach* | minix* | genix* | ultrix* | irix* \
 	     | *vms* | esix* | aix* | cnk* | sunos | sunos[34]*\
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
-	     | sym* | kopensolaris* | plan9* \
+	     | sym* | kopensolaris* | plan9* | vinix* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* \
 	     | nindy* | vxsim* | vxworks* | ebmon* | hms* | mvs* \
