diff --git libjpeg-turbo-clean/example.c libjpeg-turbo-workdir/example.c
index 78b658a..491b3f6 100644
--- libjpeg-turbo-clean/example.c
+++ libjpeg-turbo-workdir/example.c
@@ -31,6 +31,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 
 #ifdef _WIN32
 #define strcasecmp  stricmp
diff --git libjpeg-turbo-clean/md5/md5cmp.c libjpeg-turbo-workdir/md5/md5cmp.c
index 42b94ce..0c8a44f 100644
--- libjpeg-turbo-clean/md5/md5cmp.c
+++ libjpeg-turbo-workdir/md5/md5cmp.c
@@ -28,6 +28,7 @@
 
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include "./md5.h"
 #include "../tjutil.h"
 
diff --git libjpeg-turbo-clean/tjbench.c libjpeg-turbo-workdir/tjbench.c
index bedc94b..d666aec 100644
--- libjpeg-turbo-clean/tjbench.c
+++ libjpeg-turbo-workdir/tjbench.c
@@ -33,6 +33,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <ctype.h>
 #include <math.h>
 #include <errno.h>
diff --git libjpeg-turbo-clean/tjexample.c libjpeg-turbo-workdir/tjexample.c
index 1edf30f..0c67cc0 100644
--- libjpeg-turbo-clean/tjexample.c
+++ libjpeg-turbo-workdir/tjexample.c
@@ -39,6 +39,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <errno.h>
 #include <turbojpeg.h>
 
diff --git libjpeg-turbo-clean/tjunittest.c libjpeg-turbo-workdir/tjunittest.c
index dcd5006..9a033b7 100644
--- libjpeg-turbo-clean/tjunittest.c
+++ libjpeg-turbo-workdir/tjunittest.c
@@ -38,6 +38,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <limits.h>
 #include <errno.h>
 #include "tjutil.h"
diff --git libjpeg-turbo-clean/turbojpeg.c libjpeg-turbo-workdir/turbojpeg.c
index b9fe6e9..75b62c8 100644
--- libjpeg-turbo-clean/turbojpeg.c
+++ libjpeg-turbo-workdir/turbojpeg.c
@@ -38,6 +38,7 @@
 #include <jerror.h>
 #include <setjmp.h>
 #include <errno.h>
+#include <strings.h>
 #include "./turbojpeg.h"
 #include "./tjutil.h"
 #include "transupp.h"
