diff --git libjpeg-turbo-clean/src/cdjpeg.h libjpeg-turbo-workdir/src/cdjpeg.h
index dbe54a0..15a9eb3 100644
--- libjpeg-turbo-clean/src/cdjpeg.h
+++ libjpeg-turbo-workdir/src/cdjpeg.h
@@ -19,7 +19,7 @@
 #include "jpeglib.h"
 #include "jerror.h"             /* get library error codes too */
 #include "cderror.h"            /* get application-specific error codes */
-
+#include <strings.h>
 
 /*
  * Object interface for cjpeg's source file decoding modules
diff --git libjpeg-turbo-clean/src/example.c libjpeg-turbo-workdir/src/example.c
index 9405611..8a6aa61 100644
--- libjpeg-turbo-clean/src/example.c
+++ libjpeg-turbo-workdir/src/example.c
@@ -32,6 +32,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 
 #ifdef _WIN32
 #define strcasecmp  stricmp
diff --git libjpeg-turbo-clean/src/jchuff.c libjpeg-turbo-workdir/src/jchuff.c
index 8cdd5bd..84e535c 100644
--- libjpeg-turbo-clean/src/jchuff.c
+++ libjpeg-turbo-workdir/src/jchuff.c
@@ -36,7 +36,7 @@
 #endif
 #include <limits.h>
 #include "jpeg_nbits.h"
-
+#include <strings.h>
 
 /* Expanded entropy encoder object for Huffman encoding.
  *
diff --git libjpeg-turbo-clean/src/md5/md5cmp.c libjpeg-turbo-workdir/src/md5/md5cmp.c
index 42b94ce..0c8a44f 100644
--- libjpeg-turbo-clean/src/md5/md5cmp.c
+++ libjpeg-turbo-workdir/src/md5/md5cmp.c
@@ -28,6 +28,7 @@
 
 #include <stdio.h>
 #include <string.h>
+#include <strings.h>
 #include "./md5.h"
 #include "../tjutil.h"
 
diff --git libjpeg-turbo-clean/src/turbojpeg.c libjpeg-turbo-workdir/src/turbojpeg.c
index 389aea5..876a7dd 100644
--- libjpeg-turbo-clean/src/turbojpeg.c
+++ libjpeg-turbo-workdir/src/turbojpeg.c
@@ -30,6 +30,7 @@
 /* TurboJPEG/LJT:  this implements the TurboJPEG API using libjpeg or
    libjpeg-turbo */
 
+#include <strings.h>
 #include <ctype.h>
 #include <limits.h>
 #if !defined(_MSC_VER) || _MSC_VER > 1600
diff --git libjpeg-turbo-clean/src/turbojpeg.h libjpeg-turbo-workdir/src/turbojpeg.h
index 274eb7a..d97794d 100644
--- libjpeg-turbo-clean/src/turbojpeg.h
+++ libjpeg-turbo-workdir/src/turbojpeg.h
@@ -31,6 +31,7 @@
 #define __TURBOJPEG_H__
 
 #include <stddef.h>
+#include <strings.h>
 
 #if defined(_WIN32) && defined(DLLDEFINE)
 #define DLLEXPORT  __declspec(dllexport)
