From 2d9fdb5d2aee96f34ad6799d7a7f6da82b526256 Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Mon, 18 Apr 2022 05:08:43 +0200
Subject: [PATCH] Vinix specific changes

---
 configure.ac | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index de83332..72a5141 100644
--- a/configure.ac
+++ b/configure.ac
@@ -78,6 +78,7 @@ if test "$cross_compiling" = yes; then
 	    fi
             interp=
 	done
+        interp=python3
         if test x$interp = x; then
 	    AC_MSG_ERROR([python$PACKAGE_VERSION interpreter not found])
 	fi
@@ -392,6 +393,9 @@ then
 	*-*-cygwin*)
 		ac_sys_system=Cygwin
 		;;
+	*-*-vinix*)
+	    ac_sys_system=Vinix
+	    ;;
 	*-*-vxworks*)
 	    ac_sys_system=VxWorks
 	    ;;
@@ -444,6 +448,9 @@ if test "$cross_compiling" = yes; then
 	*-*-vxworks*)
 		_host_cpu=$host_cpu
 		;;
+	*-*-vinix*)
+		_host_cpu=$host_cpu
+		;;
 	*)
 		# for now, limit cross builds to known configurations
 		MACHDEP="unknown"
@@ -2673,6 +2680,9 @@ then
 	CYGWIN*)
 		LDSHARED="gcc -shared -Wl,--enable-auto-image-base"
 		LDCXXSHARED="g++ -shared -Wl,--enable-auto-image-base";;
+	Vinix*)
+		LDSHARED='$(CC) -shared'
+		LDCXXSHARED='$(CXX) -shared';;
 	*)	LDSHARED="ld";;
 	esac
 fi
@@ -2709,7 +2719,9 @@ then
 		else CCSHARED="-Kpic -belf"
 		fi;;
 	VxWorks*)
-		CCSHARED="-fpic -D__SO_PICABILINUX__  -ftls-model=global-dynamic"
+		CCSHARED="-fpic -D__SO_PICABILINUX__  -ftls-model=global-dynamic";;
+	Vinix*)
+		CCSHARED="-fPIC";;
 	esac
 fi
 AC_MSG_RESULT($CCSHARED)
@@ -2770,6 +2782,8 @@ then
 		LINKFORSHARED='-Wl,-E -N 2048K';;
 	VxWorks*)
 		LINKFORSHARED='--export-dynamic';;
+	Vinix*)
+		LINKFORSHARED='-export-dynamic';;
 	esac
 fi
 AC_MSG_RESULT($LINKFORSHARED)
-- 
2.35.3

