diff -urN --no-dereference python-clean/Modules/fcntlmodule.c python-workdir/Modules/fcntlmodule.c
--- python-clean/Modules/fcntlmodule.c	2025-04-08 15:54:08.000000000 +0200
+++ python-workdir/Modules/fcntlmodule.c	2025-04-26 16:11:44.521061432 +0200
@@ -596,8 +596,8 @@
     if (PyModule_AddIntMacro(m, F_GETOWN_EX)) return -1;
     if (PyModule_AddIntMacro(m, F_SETOWN_EX)) return -1;
     if (PyModule_AddIntMacro(m, F_OWNER_TID)) return -1;
-    if (PyModule_AddIntMacro(m, F_OWNER_PID)) return -1;
-    if (PyModule_AddIntMacro(m, F_OWNER_PGRP)) return -1;
+    //if (PyModule_AddIntMacro(m, F_OWNER_PID)) return -1;
+    //if (PyModule_AddIntMacro(m, F_OWNER_PGRP)) return -1;
 #endif
 #ifdef F_GET_RW_HINT
     // since Linux 4.13
diff -urN --no-dereference python-clean/Objects/mimalloc/prim/unix/prim.c python-workdir/Objects/mimalloc/prim/unix/prim.c
--- python-clean/Objects/mimalloc/prim/unix/prim.c	2025-04-08 15:54:08.000000000 +0200
+++ python-workdir/Objects/mimalloc/prim/unix/prim.c	2025-04-26 16:11:44.521143114 +0200
@@ -29,6 +29,10 @@
 #include <unistd.h>    // sysconf
 #include <fcntl.h>     // open, close, read, access
 
+#if defined(__vinix__)
+  #define MADV_DONTNEED 1
+#endif
+
 #if defined(__linux__)
   #include <features.h>
   #include <fcntl.h>
@@ -51,7 +55,7 @@
   #include <sys/sysctl.h>
 #endif
 
-#if !defined(__HAIKU__) && !defined(__APPLE__) && !defined(__CYGWIN__) && !defined(_AIX) && !defined(__OpenBSD__) && !defined(__FreeBSD__) && !defined(__sun) && !defined(__NetBSD__)
+#if !defined(__vinix__) && !defined(__HAIKU__) && !defined(__APPLE__) && !defined(__CYGWIN__) && !defined(_AIX) && !defined(__OpenBSD__) && !defined(__FreeBSD__) && !defined(__sun) && !defined(__NetBSD__)
   #define MI_HAS_SYSCALL_H
   #include <sys/syscall.h>
 #endif
@@ -156,6 +160,9 @@
 static int unix_madvise(void* addr, size_t size, int advice) {
   #if defined(__sun)
   return madvise((caddr_t)addr, size, advice);  // Solaris needs cast (issue #520)
+  #elif defined(__vinix__)
+  (void)addr; (void)size; (void)advice;
+  return 0;
   #else
   return madvise(addr, size, advice);
   #endif
diff -urN --no-dereference python-clean/configure python-workdir/configure
--- python-clean/configure	2025-04-08 15:54:08.000000000 +0200
+++ python-workdir/configure	2025-04-26 16:20:57.317469181 +0200
@@ -4067,6 +4067,9 @@
 	*-*-wasi*)
 	    ac_sys_system=WASI
 	    ;;
+	*-*-mlibc*)
+	    ac_sys_system=Mlibc
+	    ;;
 	*)
 		# for now, limit cross builds to known configurations
 		MACHDEP="unknown"
@@ -4491,7 +4494,7 @@
 
 if test "$cross_compiling" = yes; then
 	case "$host" in
-	*-*-linux*)
+	*-*-linux*|*-*-mlibc*)
 		case "$host_cpu" in
 		arm*)
 			_host_ident=arm
@@ -7587,7 +7590,7 @@
         PY3LIBRARY=libpython3.so
       fi
       ;;
-    Linux*|GNU*|NetBSD*|FreeBSD*|DragonFly*|OpenBSD*|VxWorks*)
+    Linux*|GNU*|NetBSD*|FreeBSD*|DragonFly*|OpenBSD*|VxWorks*|Mlibc*)
       LDLIBRARY='libpython$(LDVERSION).so'
       BLDLIBRARY='-L. -lpython$(LDVERSION)'
       RUNSHARED=LD_LIBRARY_PATH=`pwd`${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}
@@ -12983,7 +12986,7 @@
 	Emscripten*|WASI*)
 		LDSHARED='$(CC) -shared'
 		LDCXXSHARED='$(CXX) -shared';;
-	Linux*|GNU*|QNX*|VxWorks*|Haiku*)
+	Linux*|GNU*|QNX*|VxWorks*|Haiku*|Mlibc*)
 		LDSHARED='$(CC) -shared'
 		LDCXXSHARED='$(CXX) -shared';;
 	FreeBSD*)
@@ -13068,7 +13071,7 @@
 		 then CCSHARED="-fPIC";
 		 else CCSHARED="+z";
 		 fi;;
-	Linux*|GNU*) CCSHARED="-fPIC";;
+	Linux*|GNU*|Mlibc*) CCSHARED="-fPIC";;
 	Emscripten*|WASI*)
 		if test "x$enable_wasm_dynamic_linking" = xyes
 then :
@@ -13106,7 +13109,7 @@
 	    LINKFORSHARED="-Wl,-E -Wl,+s";;
 #	    LINKFORSHARED="-Wl,-E -Wl,+s -Wl,+b\$(BINLIBDEST)/lib-dynload";;
 	Linux-android*) LINKFORSHARED="-pie -Xlinker -export-dynamic";;
-	Linux*|GNU*) LINKFORSHARED="-Xlinker -export-dynamic";;
+	Linux*|GNU*|Mlibc*) LINKFORSHARED="-Xlinker -export-dynamic";;
 	# -u libsys_s pulls in all symbols in libsys
 	Darwin/*|iOS/*)
 		LINKFORSHARED="$extra_undefs -framework CoreFoundation"
