diff --git a/arch/loongarch64/bits/signal.h b/arch/loongarch64/bits/signal.h
index 1487c0d..2725276 100644
--- a/arch/loongarch64/bits/signal.h
+++ b/arch/loongarch64/bits/signal.h
@@ -32,7 +32,7 @@ struct sigcontext {
 	unsigned long sc_pc;
 	unsigned long sc_regs[32];
 	unsigned sc_flags;
-	unsigned long sc_extcontext[] __attribute__((__aligned__(16)));
+	unsigned long sc_extcontext[0] __attribute__((__aligned__(16)));
 };
 
 #define CONTEXT_INFO_ALIGN 16
@@ -82,7 +82,7 @@ typedef struct {
 	unsigned long __pc;
 	unsigned long __gregs[32];
 	unsigned __flags;
-	unsigned long __extcontext[] __attribute__((__aligned__(16)));
+	unsigned long __extcontext[0] __attribute__((__aligned__(16)));
 } mcontext_t;
 
 struct sigaltstack {
