From 42ee22d460fff5772ea40075bf2b71c829dbde78 Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Sat, 8 Jan 2022 14:37:02 +0100
Subject: [PATCH] Vinix specific changes

---
 lib/config.sub | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/config.sub b/lib/config.sub
index d74fb6d..ae4f8a3 100755
--- a/lib/config.sub
+++ b/lib/config.sub
@@ -1723,7 +1723,7 @@ case $os in
 	     | hpux* | unos* | osf* | luna* | dgux* | auroraux* | solaris* \
 	     | sym* |  plan9* | psp* | sim* | xray* | os68k* | v88r* \
 	     | hiux* | abug | nacl* | netware* | windows* \
-	     | os9* | macos* | osx* | ios* \
+	     | os9* | macos* | osx* | ios* | vinix* \
 	     | mpw* | magic* | mmixware* | mon960* | lnews* \
 	     | amigaos* | amigados* | msdos* | newsos* | unicos* | aof* \
 	     | aos* | aros* | cloudabi* | sortix* | twizzler* \
-- 
2.34.1

