diff --git gtk+-3-clean/gtk/a11y/gtkaccessibility.c gtk+-3-workdir/gtk/a11y/gtkaccessibility.c
index 7f0e520..3bf670c 100644
--- gtk+-3-clean/gtk/a11y/gtkaccessibility.c
+++ gtk+-3-workdir/gtk/a11y/gtkaccessibility.c
@@ -974,6 +974,10 @@ do_window_event_initialization (void)
 void
 _gtk_accessibility_init (void)
 {
+// We don't care about a11y at this point in time and dbus isn't working, so let's just not start that.
+#ifdef __vinix__
+  return;
+#else
   if (initialized)
     return;
 
@@ -993,4 +997,5 @@ _gtk_accessibility_init (void)
 #endif
 
   atk_misc_instance = g_object_new (GTK_TYPE_MISC_IMPL, NULL);
+#endif
 }
diff --git gtk+-3-clean/gtk/gtkmain.c gtk+-3-workdir/gtk/gtkmain.c
index ae64e18..2d43754 100644
--- gtk+-3-clean/gtk/gtkmain.c
+++ gtk+-3-workdir/gtk/gtkmain.c
@@ -355,6 +355,11 @@ static gboolean
 check_setugid (void)
 {
 /* this isn't at all relevant on MS Windows and doesn't compile ... --hb */
+#ifdef __vinix__
+  /* Vinix runs everything as root for the time being, this check is thus useless. */
+  g_warning("Vinix ignores the setugid check!\n");
+  return TRUE;
+#endif
 #ifndef G_OS_WIN32
   uid_t ruid, euid, suid; /* Real, effective and saved user ID's */
   gid_t rgid, egid, sgid; /* Real, effective and saved group ID's */
