From 5d88982bb932597b8baf1e4a2441fcc925437ea3 Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Thu, 11 Nov 2021 00:30:59 +0100
Subject: [PATCH] Vinix specific changes

---
 configure | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/configure b/configure
index 421cf85..f7570a6 100755
--- a/configure
+++ b/configure
@@ -6423,6 +6423,10 @@ CF_EOF
 		LINK_PROGS="$SHELL ${rel_builddir}/mk_prog.sh"
 		LINK_TESTS="$SHELL ${rel_builddir}/mk_prog.sh"
 		;;
+	(vinix*)
+		CC_SHARED_OPTS='-fPIC'
+		MK_SHARED_LIB='${CC} -shared -o $@'
+		;;
 	(mingw*)
 		cf_cv_shlib_version=mingw
 		cf_cv_shlib_version_infix=mingw
-- 
2.33.1

