From c9e6a12e2e262713b6d6192782105a348882f1eb Mon Sep 17 00:00:00 2001
From: mintsuki <mintsuki@protonmail.com>
Date: Mon, 6 Sep 2021 22:53:49 +0200
Subject: [PATCH] Vinix specific changes

---
 configure | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/configure b/configure
index 06f344f..e28583e 100755
--- a/configure
+++ b/configure
@@ -5961,6 +5961,10 @@ echo "${ECHO_T}$cf_cv_ldflags_search_paths_first" >&6
 
 		MK_SHARED_LIB='${CC} ${LDFLAGS} ${CFLAGS} -shared -Wl,-soname,'$cf_cv_shared_soname',-stats,-lc -o $@'
 		;;
+	(vinix*)
+		CC_SHARED_OPTS='-fPIC'
+		MK_SHARED_LIB='${CC} -shared -o $@'
+		;;
 	(mingw*)
 		cf_cv_shlib_version=mingw
 		cf_cv_shlib_version_infix=mingw
-- 
2.33.0

