diff --git ncurses-clean/c++/Makefile.in ncurses-workdir/c++/Makefile.in
index 99f21da..d0c6eb0 100644
--- ncurses-clean/c++/Makefile.in
+++ ncurses-workdir/c++/Makefile.in
@@ -182,7 +182,7 @@ AUTO_SRC	= \
 all \
 libs ::		$(AUTO_SRC) $(LIBRARIES)
 
-@MAKE_TESTS@all ::		demo$x
+#@MAKE_TESTS@all ::		demo$x
 
 depend :
 
@@ -256,20 +256,20 @@ $(INCDIR)/panel.h :
 
 ###############################################################################
 
-OBJS_DEMO = $(MODEL)/demo$o
+#OBJS_DEMO = $(MODEL)/demo$o
 
-$(MODEL)/demo$o : $(srcdir)/demo.cc \
-		$(HEADER_DEPS) \
-		$(cursesf_h) \
-		$(cursesm_h) \
-		$(cursesapp_h)
-	@echo 'compiling demo (obj_s)'
-	@( cd $(MODEL) && $(LIBTOOL_COMPILE) $(CXX) $(CFLAGS_DEFAULT) -c $(top_srcdir)/c++/demo.cc -o $@ )
+#$(MODEL)/demo$o : $(srcdir)/demo.cc \
+#		$(HEADER_DEPS) \
+#		$(cursesf_h) \
+#		$(cursesm_h) \
+#		$(cursesapp_h)
+#	@echo 'compiling demo (obj_s)'
+#	@( cd $(MODEL) && $(LIBTOOL_COMPILE) $(CXX) $(CFLAGS_DEFAULT) -c $(top_srcdir)/c++/demo.cc -o $@ )
 
-demo$x:	$(OBJS_DEMO) \
-	$(LIBRARIES)  \
-	@TEST_DEPS@
-	@ECHO_LD@ $(LINK) -o $@ $(OBJS_DEMO) $(LDFLAGS_DEFAULT)
+#demo$x:	$(OBJS_DEMO) \
+#	$(LIBRARIES)  \
+#	@TEST_DEPS@
+#	@ECHO_LD@ $(LINK) -o $@ $(OBJS_DEMO) $(LDFLAGS_DEFAULT)
 
 etip.h:	$(srcdir)/etip.h.in $(srcdir)/edit_cfg.sh
 	cp $(srcdir)/etip.h.in $@
diff --git ncurses-clean/configure ncurses-workdir/configure
index 814eb2f..3e2ae5c 100755
--- ncurses-clean/configure
+++ ncurses-workdir/configure
@@ -6925,6 +6925,10 @@ CF_EOF
 		LINK_PROGS="$SHELL ${rel_builddir}/mk_prog.sh"
 		LINK_TESTS="$SHELL ${rel_builddir}/mk_prog.sh"
 		;;
+	(vinix*)
+		CC_SHARED_OPTS='-fPIC'
+		MK_SHARED_LIB='${CC} -shared -o $@'
+		;;
 	(mingw*)
 		cf_cv_shlib_version=mingw
 		cf_cv_shlib_version_infix=mingw
