diff -urN --no-dereference ncurses-clean/c++/Makefile.in ncurses-workdir/c++/Makefile.in
--- ncurses-clean/c++/Makefile.in
+++ ncurses-workdir/c++/Makefile.in
@@ -120,7 +120,7 @@
 		-l@FORM_NAME@@USE_LIB_SUFFIX@ \
 		-l@MENU_NAME@@USE_LIB_SUFFIX@ \
 		-l@PANEL_NAME@@USE_LIB_SUFFIX@ \
-		-lncurses@USE_LIB_SUFFIX@ @SHLIB_LIST@
+		-lncurses@USE_LIB_SUFFIX@
 
 LIBROOT		= ncurses++
 
@@ -159,7 +159,7 @@
 LDFLAGS_DEFAULT	= $(LINK_@DFT_UPR_MODEL@) $(LDFLAGS_@DFT_UPR_MODEL@)
 
 # flags for library built by this makefile
-LDFLAGS		= $(TEST_ARGS) @LDFLAGS@ \
+LDFLAGS		= @LDFLAGS@ \
 	@LD_MODEL@ $(TEST_LIBS) @LIBS@ $(CXXLIBS)
 
 AUTO_SRC	= \
@@ -184,7 +184,7 @@
 all \
 libs ::		$(AUTO_SRC) $(LIBRARIES)
 
-@MAKE_TESTS@all ::		demo$x
+#@MAKE_TESTS@all ::		demo$x
 
 depend :
 
@@ -258,20 +258,20 @@
 
 ###############################################################################
 
-OBJS_DEMO = $(MODEL)/demo$o
+#OBJS_DEMO = $(MODEL)/demo$o
 
-$(MODEL)/demo$o : $(srcdir)/demo.cc \
-		$(HEADER_DEPS) \
-		$(cursesf_h) \
-		$(cursesm_h) \
-		$(cursesapp_h)
-	@echo 'compiling demo (obj_s)'
-	@( cd $(MODEL) && $(LIBTOOL_COMPILE) $(CXX) $(CFLAGS_DEFAULT) -c $(top_srcdir)/c++/demo.cc -o $@ )
-
-demo$x:	$(OBJS_DEMO) \
-	$(LIBRARIES)  \
-	@TEST_DEPS@
-	@ECHO_LD@ $(LINK) -o $@ $(OBJS_DEMO) $(LDFLAGS_DEFAULT)
+#$(MODEL)/demo$o : $(srcdir)/demo.cc \
+#		$(HEADER_DEPS) \
+#		$(cursesf_h) \
+#		$(cursesm_h) \
+#		$(cursesapp_h)
+#	@echo 'compiling demo (obj_s)'
+#	@( cd $(MODEL) && $(LIBTOOL_COMPILE) $(CXX) $(CFLAGS_DEFAULT) -c $(top_srcdir)/c++/demo.cc -o $@ )
+
+#demo$x:	$(OBJS_DEMO) \
+#	$(LIBRARIES)  \
+#	@TEST_DEPS@
+#	@ECHO_LD@ $(LINK) -o $@ $(OBJS_DEMO) $(LDFLAGS_DEFAULT)
 
 etip.h:	$(srcdir)/etip.h.in $(srcdir)/edit_cfg.sh
 	cp $(srcdir)/etip.h.in $@
diff -urN --no-dereference ncurses-clean/configure ncurses-workdir/configure
--- ncurses-clean/configure
+++ ncurses-workdir/configure
@@ -3821,7 +3821,7 @@
 
 		save_CPPFLAGS="$CPPFLAGS"
 		eval cf_includedir=${includedir}
-		CPPFLAGS="$CPPFLAGS -I${cf_includedir}"
+		CPPFLAGS="$CPPFLAGS"
 
 cat >"conftest.$ac_ext" <<_ACEOF
 #line 3827 "configure"
@@ -7270,6 +7270,10 @@
 		LINK_PROGS="$SHELL ${rel_builddir}/mk_prog.sh"
 		LINK_TESTS="$SHELL ${rel_builddir}/mk_prog.sh"
 		;;
+	(vinix*)
+		CC_SHARED_OPTS='-fPIC'
+		MK_SHARED_LIB='${CC} -shared -o $@'
+		;;
 	(mingw*)
 		cf_cv_shlib_version=mingw
 		cf_cv_shlib_version_infix=mingw
diff -urN --no-dereference ncurses-clean/form/Makefile.in ncurses-workdir/form/Makefile.in
--- ncurses-clean/form/Makefile.in
+++ ncurses-workdir/form/Makefile.in
@@ -112,7 +112,7 @@
 LDFLAGS		= @LDFLAGS@ @LD_MODEL@ @LIBS@
 
 SHLIB_DIRS	= -L../lib
-SHLIB_LIST	= $(SHLIB_DIRS) -lncurses@USE_LIB_SUFFIX@ @SHLIB_LIST@
+SHLIB_LIST	= $(SHLIB_DIRS) -lncurses@USE_LIB_SUFFIX@
 
 RPATH_LIST	= @RPATH_LIST@
 RESULTING_SYMS	= @RESULTING_SYMS@
diff -urN --no-dereference ncurses-clean/misc/gen-pkgconfig.in ncurses-workdir/misc/gen-pkgconfig.in
--- ncurses-clean/misc/gen-pkgconfig.in
+++ ncurses-workdir/misc/gen-pkgconfig.in
@@ -83,7 +83,7 @@
 fi
 
 lib_flags=
-for opt in -L$libdir @EXTRA_PKG_LDFLAGS@ @LIBS@
+for opt in -L$libdir @LIBS@
 do
 	case $opt in
 	-l*) # LIBS is handled specially below
diff -urN --no-dereference ncurses-clean/misc/ncurses-config.in ncurses-workdir/misc/ncurses-config.in
--- ncurses-clean/misc/ncurses-config.in
+++ ncurses-workdir/misc/ncurses-config.in
@@ -110,7 +110,7 @@
 # There is no portable way to find the list of standard library directories.
 # Require a POSIX shell anyway, to keep this simple.
 lib_flags=
-for opt in -L$libdir @EXTRA_PKG_LDFLAGS@ $LIBS
+for opt in -L$libdir $LIBS
 do
 	case $opt in
 	-specs*) # ignore linker specs-files which were used to build library
diff -urN --no-dereference ncurses-clean/panel/Makefile.in ncurses-workdir/panel/Makefile.in
--- ncurses-clean/panel/Makefile.in
+++ ncurses-workdir/panel/Makefile.in
@@ -114,7 +114,7 @@
 LDFLAGS		= @LDFLAGS@ @LD_MODEL@ @LIBS@
 
 SHLIB_DIRS	= -L../lib
-SHLIB_LIST	= $(SHLIB_DIRS) -lncurses@USE_LIB_SUFFIX@ @SHLIB_LIST@
+SHLIB_LIST	= $(SHLIB_DIRS) -lncurses@USE_LIB_SUFFIX@
 
 RPATH_LIST	= @RPATH_LIST@
 RESULTING_SYMS	= @RESULTING_SYMS@
diff -urN --no-dereference ncurses-clean/progs/Makefile.in ncurses-workdir/progs/Makefile.in
--- ncurses-clean/progs/Makefile.in
+++ ncurses-workdir/progs/Makefile.in
@@ -80,7 +80,7 @@
 LIBTOOL_UNINSTALL = @LIB_UNINSTALL@
 
 INSTALL		= @INSTALL@
-INSTALL_PROG	= @INSTALL_PROGRAM@ @INSTALL_OPT_S@
+INSTALL_PROG	= @INSTALL_PROGRAM@
 transform	= @program_transform_name@
 
 AWK		= @AWK@
