diff --git libffi-clean/configure.ac libffi-workdir/configure.ac
index 7e8cd98..86b4b49 100644
--- libffi-clean/configure.ac
+++ libffi-workdir/configure.ac
@@ -1,6 +1,6 @@
 dnl Process this with autoconf to create configure
 
-AC_PREREQ([2.71])
+AC_PREREQ([2.69])
 
 AC_INIT([libffi],[3.4.4],[http://github.com/libffi/libffi/issues])
 AC_CONFIG_HEADERS([fficonfig.h])
@@ -95,7 +95,7 @@ m4_warn([obsolete],
   ISO C90 headers unconditionally.])dnl
 # Autoupdate added the next two lines to ensure that your configure
 # script's behavior did not change.  They are probably safe to remove.
-AC_CHECK_INCLUDES_DEFAULT
+#AC_CHECK_INCLUDES_DEFAULT
 AC_PROG_EGREP
 
 AC_CHECK_FUNCS(memcpy)
