diff --git mesa-demos-clean/meson.build mesa-demos-workdir/meson.build
index c1853f9..ea9d9f0 100644
--- mesa-demos-clean/meson.build
+++ mesa-demos-workdir/meson.build
@@ -99,7 +99,7 @@ if not dep_glu.found() and host_machine.system() != 'darwin'
 endif
 
 dep_glx = dependency('glx', required: false, disabler : true)
-if not dep_glx.found() and host_machine.system() == 'darwin'
+if not dep_glx.found() and (host_machine.system() == 'darwin' or host_machine.system() == 'vinix')
   # xquartz doesn't have a glx.pc, but it does have the header. And all the
   # symbols reside in libGL, so let's just use that.
   if cc.check_header('GL/glx.h', dependencies: dep_x11)
