diff --git openssh-clean/configure.ac openssh-workdir/configure.ac
index 591d5a3..0a0e5cc 100644
--- openssh-clean/configure.ac
+++ openssh-workdir/configure.ac
@@ -719,6 +719,10 @@ case "$host" in
 	AC_DEFINE([DISABLE_UTMP], [1], [Define if you don't want to use utmp])
 	AC_DEFINE([DISABLE_WTMP], [1], [Define if you don't want to use wtmp])
 	;;
+*-*-vinix*-*)
+	AC_DEFINE([NO_UID_RESTORATION_TEST], [1],
+		[Define to disable UID restoration test])
+	;;
 *-*-cygwin*)
 	LIBS="$LIBS /usr/lib/textreadmode.o"
 	AC_DEFINE([HAVE_CYGWIN], [1], [Define if you are on Cygwin])
